<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>Background &mdash; TRAL 0.3.5 documentation</title>
    
    <link rel="stylesheet" href="_static/tral.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    './',
        VERSION:     '0.3.5',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <link rel="top" title="TRAL 0.3.5 documentation" href="index.html" />
    <link rel="next" title="Annotate tandem repeats from sequence domain models." href="cphmm.html" />
    <link rel="prev" title="Run and parse de novo repeat detection software." href="denovo.html" />
    <link href='http://fonts.googleapis.com/css?family=Source+Sans+Pro:300,400,700|Source+Code+Pro|Armata|IM+Fell+English'
          rel='stylesheet' type='text/css' />
 
    <style type="text/css">
      table.right { float: right; margin-left: 20px; }
      table.right td { border: 1px solid #ccc; }
      
    </style>
    <script type="text/javascript">
      // intelligent scrolling of the sidebar content
      $(window).scroll(function() {
        var sb = $('.sphinxsidebarwrapper');
        var win = $(window);
        var sbh = sb.height();
        var offset = $('.sphinxsidebar').position()['top'];
        var wintop = win.scrollTop();
        var winbot = wintop + win.innerHeight();
        var curtop = sb.position()['top'];
        var curbot = curtop + sbh;
        // does sidebar fit in window?
        if (sbh < win.innerHeight()) {
          // yes: easy case -- always keep at the top
          sb.css('top', $u.min([$u.max([0, wintop - offset - 10]),
                                $(document).height() - sbh - 200]));
        } else {
          // no: only scroll if top/bottom edge of sidebar is at
          // top/bottom edge of window
          if (curtop > wintop && curbot > winbot) {
            sb.css('top', $u.max([wintop - offset - 10, 0]));
          } else if (curtop < wintop && curbot < winbot) {
            sb.css('top', $u.min([winbot - sbh - offset - 20,
                                  $(document).height() - sbh - 200]));
          }
        }
      });
    </script>

  </head>
  <body role="document">
<div class="pageheader">
    <a href="http://www.isb-sib.ch/" class="sib_logo" title="SIB Swiss Institute of Bioinformatics"><img src="_static/sib.png" alt="SIB" class="sib_logo" /></a>
    <a href="index.html">
      <img class="headlogo" src="_static/trallogo.png" alt="TRAL" />
      <span class="headlogo">Tandem Repeat Annotation Library</span>
    </a>
    <div class="headlinks">
        <a href="index.html">Home</a>
        <a href="install.html">Install</a>
        <a href="code_docs.html">Code docs</a>
    </div>
</div>


      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
  <h3><a href="index.html">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">Background</a><ul>
<li><a class="reference internal" href="#more-tandem-repeat-characteristics">More tandem repeat characteristics</a></li>
<li><a class="reference internal" href="#writing-a-tandem-repeat-to-csv">Writing a tandem repeat to .csv</a></li>
</ul>
</li>
</ul>

  <h4>Previous topic</h4>
  <p class="topless"><a href="denovo.html"
                        title="previous chapter">Run and parse de novo repeat detection software.</a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="cphmm.html"
                        title="next chapter">Annotate tandem repeats from sequence domain models.</a></p>
  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="_sources/background.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="section" id="background">
<span id="id1"></span><h1>Background<a class="headerlink" href="#background" title="Permalink to this headline">¶</a></h1>
<p>When printing a tandem repeat, you will see a lot of information. Here is what it means.
(You get this example from <a class="reference internal" href="significance_test.html#significance-test"><span>the significance test tutorials</span></a>)</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="k">print</span><span class="p">(</span><span class="n">lHIV_Sequence</span><span class="p">[</span><span class="mi">4</span><span class="p">]</span><span class="o">.</span><span class="n">dRepeat_list</span><span class="p">[</span><span class="s1">&#39;denovo&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">repeats</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>
<span class="go">&gt; begin:38 l_effective:2 n:6 pvalue:0.0 divergence:0.46649169922545164 type:phylo_gap01</span>
<span class="go">RRN</span>
<span class="go">RR-</span>
<span class="go">RRW</span>
<span class="go">RA-</span>
<span class="go">RQ-</span>
<span class="go">RQI</span>
</pre></div>
</div>
<p>The tandem repeat is displayed as an alignment of tandem repeat units, similar to a
<a class="reference external" href="http://en.wikipedia.org/wiki/Multiple_sequence_alignment">multiple sequence alignment</a>:</p>
<div class="highlight-python"><div class="highlight"><pre>RRN
RR-
RRW
RA-
RQ-
RQI
</pre></div>
</div>
<p>Within the sequence, this tandem repeat would like the following:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="n">RRNRRRRWRARQRQI</span>
</pre></div>
</div>
<p>These are the other characteristics of the tandem repeat, that can be shown (if available):</p>
<ul class="simple">
<li><strong>begin</strong>: Index where the tandem repeat starts in the sequence</li>
<li><strong>l_effective</strong>: Length of the consensus tandem repeat unit, not including insertions.</li>
<li><strong>n</strong>: Number of tandem repeat units.</li>
<li><strong>pvalue</strong>: What is the probability that this repeat has occurred by random chance? (This value is only a good estimate. So even if pvalue=0.0, of course it is possible that the sequence shows similarity by random chance, and not because the repeat units have evolved by duplications).</li>
<li><strong>divergence</strong>:  Model-based measure of the similarity of the repeat units.
(mathematically, it is the maximum likelihood estimate of the branch length on the phylogeny
connecting all tandem repeat units. In the above example, every site has mutated 0.47 times
on average. A value of 0 would mean no mutations have occurred, and the sequence of the
tandem repeat units is very conserved).</li>
<li><strong>type</strong>: The model which was used to calculate the pvalue and the divergence.</li>
</ul>
<div class="section" id="more-tandem-repeat-characteristics">
<h2>More tandem repeat characteristics<a class="headerlink" href="#more-tandem-repeat-characteristics" title="Permalink to this headline">¶</a></h2>
<p>You have access to more tandem repeat characteristics. The dir() command will provide you
with a list of all attributes and functions connected to the object:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="nb">dir</span><span class="p">(</span><span class="n">lHIV_Sequence</span><span class="p">[</span><span class="mi">4</span><span class="p">]</span><span class="o">.</span><span class="n">dRepeat_list</span><span class="p">[</span><span class="s1">&#39;denovo&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">repeats</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>
<span class="go">[&#39;TRD&#39;, &#39;__class__&#39;, &#39;__delattr__&#39;, &#39;__dict__&#39;, &#39;__dir__&#39;, &#39;__doc__&#39;, &#39;__eq__&#39;, &#39;__format__&#39;, &#39;__ge__&#39;, &#39;__getattribute__&#39;, &#39;__gt__&#39;, &#39;__hash__&#39;, &#39;__init__&#39;, &#39;__le__&#39;, &#39;__lt__&#39;, &#39;__module__&#39;, &#39;__ne__&#39;, &#39;__new__&#39;, &#39;__reduce__&#39;, &#39;__reduce_ex__&#39;, &#39;__repr__&#39;, &#39;__setattr__&#39;, &#39;__sizeof__&#39;, &#39;__str__&#39;, &#39;__subclasshook__&#39;, &#39;__weakref__&#39;, &#39;begin&#39;, &#39;calc_index_msa&#39;, &#39;calc_calc_n_effective&#39;, &#39;calculate_pvalues&#39;, &#39;calculate_scores&#39;, &#39;create&#39;, &#39;dDivergence&#39;, &#39;dPValue&#39;, &#39;dScore&#39;, &#39;deleteInsertionColumns&#39;, &#39;deletions&#39;, &#39;divergence&#39;, &#39;gapStructure&#39;, &#39;gap_structure_HMM&#39;, &#39;gaps&#39;, &#39;insertions&#39;, &#39;l&#39;, &#39;l_effective&#39;, &#39;msa&#39;, &#39;msaD&#39;, &#39;msaT&#39;, &#39;msaTD&#39;, &#39;msaTDN&#39;, &#39;msaTD_standard_aa&#39;, &#39;msa_original&#39;, &#39;msa_standard_aa&#39;, &#39;n&#39;, &#39;calc_n_effective&#39;, &#39;nGap&#39;, &#39;pvalue&#39;, &#39;save_original_msa&#39;, &#39;score&#39;, &#39;sequence_length&#39;, &#39;sequence_type&#39;, &#39;text&#39;, &#39;textD&#39;, &#39;textD_standard_aa&#39;, &#39;totD&#39;, &#39;write&#39;]</span>
</pre></div>
</div>
<p>For a more explanation, check out the <a class="reference internal" href="code_docs.html#code-docs"><span>Class documentation</span></a>, or
<a class="reference internal" href="contributors.html#contributors"><span>contact us</span></a>.</p>
</div>
<div class="section" id="writing-a-tandem-repeat-to-csv">
<h2>Writing a tandem repeat to .csv<a class="headerlink" href="#writing-a-tandem-repeat-to-csv" title="Permalink to this headline">¶</a></h2>
<p>When you write a tandem repeat to .csv with TRAL, the result may look as follows:</p>
<div class="highlight-python"><div class="highlight"><pre>begin   msa_original    l_effective      n_effective      repeat_region_length divergence      pvalue
316     GDII,GDIR       4       2.0     8       None    None
507     FLG,FLG 3       2.0     6       None    None
</pre></div>
</div>
<p>Additional to the tandem repeat characteristics explained above, here you can find:</p>
<ul class="simple">
<li><strong>msa_original</strong>: The tandem repeat unit alignment, with units separated by commata.</li>
<li><strong>repeat_region_length</strong>: The number of characters covered by the tandem repeat region.</li>
</ul>
<p><em>None</em> values indicated that the required characteristics had not been calculated previously
in the code.</p>
</div>
</div>


          </div>
        </div>
      </div>
      <div class="clearer"></div>
    </div>


  </body>
</html>