
<!DOCTYPE html>

<html lang="en">
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="generator" content="Docutils 0.19: https://docutils.sourceforge.io/" />

    <title>Configuration &#8212; TRAL 2.0.0 documentation</title>
    <link rel="stylesheet" type="text/css" href="_static/pygments.css" />
    <link rel="stylesheet" type="text/css" href="_static/tral.css" />
    <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
    <script src="_static/doctools.js"></script>
    <script src="_static/sphinx_highlight.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Using TRAL in a container or VM" href="install_containerized.html" />
    <link rel="prev" title="p-Value distribution files" href="pvaluefiles.html" />
    <link href='http://fonts.googleapis.com/css?family=Source+Sans+Pro:300,400,700|Source+Code+Pro|Armata|IM+Fell+English'
          rel='stylesheet' type='text/css' />
 
    <style type="text/css">
      table.right { float: right; margin-left: 20px; }
      table.right td { border: 1px solid #ccc; }
      
    </style>
    <script type="text/javascript">
      // intelligent scrolling of the sidebar content
      $(window).scroll(function() {
        var sb = $('.sphinxsidebarwrapper');
        var win = $(window);
        var sbh = sb.height();
        var offset = $('.sphinxsidebar').position()['top'];
        var wintop = win.scrollTop();
        var winbot = wintop + win.innerHeight();
        var curtop = sb.position()['top'];
        var curbot = curtop + sbh;
        // does sidebar fit in window?
        if (sbh < win.innerHeight()) {
          // yes: easy case -- always keep at the top
          sb.css('top', $u.min([$u.max([0, wintop - offset - 10]),
                                $(document).height() - sbh - 200]));
        } else {
          // no: only scroll if top/bottom edge of sidebar is at
          // top/bottom edge of window
          if (curtop > wintop && curbot > winbot) {
            sb.css('top', $u.max([wintop - offset - 10, 0]));
          } else if (curtop < wintop && curbot < winbot) {
            sb.css('top', $u.min([winbot - sbh - offset - 20,
                                  $(document).height() - sbh - 200]));
          }
        }
      });
    </script>

  </head><body>
<div class="pageheader">
    <a href="http://www.isb-sib.ch/" class="sib_logo" title="SIB Swiss Institute of Bioinformatics"><img src="_static/sib.png" alt="SIB" class="sib_logo" /></a>
    <a href="index.html">
      <img class="headlogo" src="_static/trallogo.png" alt="TRAL" />
      <span class="headlogo">Tandem Repeat Annotation Library</span>
    </a>
    <div class="headlinks">
        <a href="index.html">Home</a>
        <a href="install.html">Install</a>
        <a href="code_docs.html">Code docs</a>
    </div>
</div>


      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
  <div>
    <h3><a href="index.html">Table of Contents</a></h3>
    <ul>
<li><a class="reference internal" href="#">Configuration</a><ul>
<li><a class="reference internal" href="#configs-ini-required-supply-paths-to-external-software-and-data">configs.ini - REQUIRED: Supply paths to external software and data.</a><ul>
<li><a class="reference internal" href="#de-novo-repeat-detection"><em>De novo</em> repeat detection</a></li>
<li><a class="reference internal" href="#build-hmmer-models">Build Hmmer models</a></li>
<li><a class="reference internal" href="#realign-tandem-repeat-units">Realign tandem repeat units</a></li>
<li><a class="reference internal" href="#simulation-of-evolution-in-tandem-repeats">Simulation of evolution in tandem repeats</a></li>
</ul>
</li>
<li><a class="reference internal" href="#configs-ini-optional-change-tral-s-default-behaviour">configs.ini - OPTIONAL: Change TRAL’s default behaviour</a><ul>
<li><a class="reference internal" href="#define-the-default-filter-behaviour">Define the default filter behaviour.</a></li>
<li><a class="reference internal" href="#creation-of-new-repeat-instances">Creation of new Repeat instances</a></li>
<li><a class="reference internal" href="#statistical-significance-calculation-and-models-of-repeat-evolution">Statistical significance calculation and models of repeat evolution</a></li>
<li><a class="reference internal" href="#creation-of-new-repeat-list-instances">Creation of new Repeat list instances</a></li>
<li><a class="reference internal" href="#restrict-hmmer-model-size">Restrict Hmmer model size.</a></li>
</ul>
</li>
<li><a class="reference internal" href="#logging-ini-optional">logging.ini - OPTIONAL</a></li>
</ul>
</li>
</ul>

  </div>
  <div>
    <h4>Previous topic</h4>
    <p class="topless"><a href="pvaluefiles.html"
                          title="previous chapter">p-Value distribution files</a></p>
  </div>
  <div>
    <h4>Next topic</h4>
    <p class="topless"><a href="install_containerized.html"
                          title="next chapter">Using TRAL in a container or VM</a></p>
  </div>
  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="_sources/configure.rst.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div>
<div id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" autocomplete="off" autocorrect="off" autocapitalize="off" spellcheck="false"/>
      <input type="submit" value="Go" />
    </form>
    </div>
</div>
<script>document.getElementById('searchbox').style.display = "block"</script>
        </div>
      </div>

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <section id="configuration">
<span id="configure"></span><h1>Configuration<a class="headerlink" href="#configuration" title="Permalink to this heading">¶</a></h1>
<p>Here you see how to configure TRAL after installation. All configuration files are located
in your home directory:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="o">~/.</span><span class="n">tral</span><span class="o">/</span>
</pre></div>
</div>
<section id="configs-ini-required-supply-paths-to-external-software-and-data">
<h2>configs.ini - REQUIRED: Supply paths to external software and data.<a class="headerlink" href="#configs-ini-required-supply-paths-to-external-software-and-data" title="Permalink to this heading">¶</a></h2>
<p>In configs.ini you can set the paths to all external software and data used in TRAL, as well
as change default thresholds and behaviours.</p>
<section id="de-novo-repeat-detection">
<h3><em>De novo</em> repeat detection<a class="headerlink" href="#de-novo-repeat-detection" title="Permalink to this heading">¶</a></h3>
<p>Enlist all <em>de novo</em> detection algorithms you wish to use. All other algorithms are ignored
by default.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">sequence</span><span class="p">]</span>
    <span class="p">[[</span><span class="n">repeat_detection</span><span class="p">]]</span>
        <span class="n">AA</span> <span class="o">=</span> <span class="n">HHrepID</span><span class="p">,</span> <span class="n">T</span><span class="o">-</span><span class="n">REKS</span><span class="p">,</span> <span class="n">TRUST</span><span class="p">,</span> <span class="n">XSTREAM</span>  <span class="c1"># List of default de novo detectors on amino acid data that you wish to use.</span>
        <span class="n">DNA</span> <span class="o">=</span> <span class="n">PHOBOS</span><span class="p">,</span> <span class="n">XSTREAM</span>  <span class="c1">#List of default de novo detectors on nucleic data that you wish to use.</span>
</pre></div>
</div>
<p>If the binaries or executable scripts of any <em>de novo</em> detection algorithm are not in the
system path, set the absolute paths in config.ini. Details for each detector (e.g. to create
executable scripts) are detailed in <a class="reference internal" href="install_external.html#install-external"><span class="std std-ref">Installation of external software</span></a>.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">sequence</span><span class="p">]</span>
    <span class="p">[[</span><span class="n">repeat_detector_path</span><span class="p">]]</span>
        <span class="n">PHOBOS</span> <span class="o">=</span> <span class="n">phobos</span>
        <span class="n">HHrepID</span> <span class="o">=</span> <span class="n">hhrepid_64</span>
        <span class="n">HHrepID_dummyhmm</span> <span class="o">=</span> <span class="o">/</span><span class="n">path</span><span class="o">/</span><span class="n">to</span><span class="o">/</span><span class="n">home</span><span class="o">/.</span><span class="n">tral</span><span class="o">/</span><span class="n">data</span><span class="o">/</span><span class="n">hhrepid</span><span class="o">/</span><span class="n">dummyHMM</span><span class="o">.</span><span class="n">hmm</span>
        <span class="n">TRUST</span> <span class="o">=</span> <span class="n">TRUST</span>
        <span class="n">TRUST_substitutionmatrix</span> <span class="o">=</span> <span class="o">/</span><span class="n">path</span><span class="o">/</span><span class="n">to</span><span class="o">/</span><span class="n">TRUST</span><span class="o">/</span><span class="n">Align</span><span class="o">/</span><span class="n">BLOSUM50</span>  <span class="c1">#Recommended: Use matrices packages shipped with TRUST</span>
        <span class="n">XSTREAM</span> <span class="o">=</span> <span class="n">XSTREAM</span>
        <span class="p">[</span><span class="o">...</span><span class="p">]</span>
</pre></div>
</div>
</section>
<section id="build-hmmer-models">
<h3>Build Hmmer models<a class="headerlink" href="#build-hmmer-models" title="Permalink to this heading">¶</a></h3>
<p>If <em>hmmbuild</em> is not in your system path, define the absolute path:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">hmm</span><span class="p">]</span>
    <span class="n">hmmbuild</span> <span class="o">=</span> <span class="n">path</span><span class="o">/</span><span class="n">to</span><span class="o">/</span><span class="n">hmmbuild</span>
</pre></div>
</div>
</section>
<section id="realign-tandem-repeat-units">
<h3>Realign tandem repeat units<a class="headerlink" href="#realign-tandem-repeat-units" title="Permalink to this heading">¶</a></h3>
<p>At current, tandem repeat units are realigned with Mafft’s ginsi global aligner. This is
required for example if you annotate tandem repeats from profile models. If <em>ginsi</em> is
not in your system path, define the absolute path:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">repeat</span><span class="p">]</span>
    <span class="n">ginsi</span> <span class="o">=</span> <span class="o">/</span><span class="n">path</span><span class="o">/</span><span class="n">to</span><span class="o">/</span><span class="n">ginsi</span>
</pre></div>
</div>
<p>Additionally, all detected tandem repeats (either from profile models or <em>denovo</em> can be realigned with the indel-aware proPIP alignment algorithm.
To use proPIP you need to have Castor (incl. aligner) which ships with proPIP. If <em>Castor</em> is not in your system path, define the absolute path:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">repeat</span><span class="p">]</span>
    <span class="n">Castor</span> <span class="o">=</span> <span class="o">/</span><span class="n">path</span><span class="o">/</span><span class="n">to</span><span class="o">/</span><span class="n">Castor</span>
    <span class="p">[[</span><span class="n">castor_parameter</span><span class="p">]]</span>
        <span class="n">rate_distribution</span> <span class="o">=</span> <span class="n">constant</span> <span class="c1"># either constant or gamma</span>
</pre></div>
</div>
</section>
<section id="simulation-of-evolution-in-tandem-repeats">
<h3>Simulation of evolution in tandem repeats<a class="headerlink" href="#simulation-of-evolution-in-tandem-repeats" title="Permalink to this heading">¶</a></h3>
<p>If <em>alfsim</em> is not in your system path, set the absolute path:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">repeat</span><span class="p">]</span>
        <span class="n">alfsim</span> <span class="o">=</span> <span class="n">path</span><span class="o">/</span><span class="n">to</span><span class="o">/</span><span class="n">alfsim</span>
</pre></div>
</div>
</section>
</section>
<section id="configs-ini-optional-change-tral-s-default-behaviour">
<h2>configs.ini - OPTIONAL: Change TRAL’s default behaviour<a class="headerlink" href="#configs-ini-optional-change-tral-s-default-behaviour" title="Permalink to this heading">¶</a></h2>
<section id="define-the-default-filter-behaviour">
<h3>Define the default filter behaviour.<a class="headerlink" href="#define-the-default-filter-behaviour" title="Permalink to this heading">¶</a></h3>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="nb">filter</span><span class="p">]</span>
    <span class="p">[[</span><span class="n">basic</span><span class="p">]]</span>
        <span class="n">tag</span> <span class="o">=</span> <span class="n">basic_filter</span>
        <span class="p">[[[</span><span class="nb">dict</span><span class="p">]]]</span>
            <span class="p">[[[[</span><span class="n">pvalue</span><span class="p">]]]]</span>
                <span class="n">func_name</span> <span class="o">=</span> <span class="n">pvalue</span>
                <span class="n">score</span> <span class="o">=</span> <span class="n">phylo_gap01</span>  <span class="c1"># score</span>
                <span class="n">threshold</span> <span class="o">=</span> <span class="mf">0.1</span>  <span class="c1"># p-Value cut-off</span>
            <span class="p">[[[[</span><span class="n">n_effective</span><span class="p">]]]]</span>
                <span class="n">func_name</span> <span class="o">=</span> <span class="n">attribute</span>
                <span class="n">attribute</span> <span class="o">=</span> <span class="n">n_effective</span>  <span class="c1"># attribute of tandem repeat, e.g. the repeat unit length l_effective</span>
                <span class="nb">type</span> <span class="o">=</span> <span class="nb">min</span>
                <span class="n">threshold</span> <span class="o">=</span> <span class="mf">1.9</span>  <span class="c1"># p-Value cut-off</span>
</pre></div>
</div>
</section>
<section id="creation-of-new-repeat-instances">
<h3>Creation of new Repeat instances<a class="headerlink" href="#creation-of-new-repeat-instances" title="Permalink to this heading">¶</a></h3>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">repeat</span><span class="p">]</span>
    <span class="n">scoreslist</span> <span class="o">=</span> <span class="n">phylo_gap01</span><span class="p">,</span>  <span class="c1"># score</span>
    <span class="n">calc_score</span> <span class="o">=</span> <span class="kc">False</span>  <span class="c1"># Is the score calculated?</span>
    <span class="n">calc_pvalue</span> <span class="o">=</span> <span class="kc">False</span> <span class="c1"># Is the pvalue calculated?</span>
    <span class="n">precision</span> <span class="o">=</span> <span class="mi">10</span>
    <span class="n">ginsi</span> <span class="o">=</span> <span class="o">/</span><span class="n">path</span><span class="o">/</span><span class="n">to</span><span class="o">/</span><span class="n">ginsi</span>  <span class="c1"># Path to the mafft global aligner ginsi.</span>
    <span class="n">Castor</span> <span class="o">=</span> <span class="n">Castor</span>
    <span class="p">[[</span><span class="n">castor_parameter</span><span class="p">]]</span>
        <span class="n">rate_distribution</span> <span class="o">=</span> <span class="n">constant</span> <span class="c1"># either constant or gamma</span>
    <span class="n">alfsim</span> <span class="o">=</span> <span class="n">alfsim</span>
</pre></div>
</div>
</section>
<section id="statistical-significance-calculation-and-models-of-repeat-evolution">
<h3>Statistical significance calculation and models of repeat evolution<a class="headerlink" href="#statistical-significance-calculation-and-models-of-repeat-evolution" title="Permalink to this heading">¶</a></h3>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">repeat_score</span><span class="p">]</span>
    <span class="n">evolutionary_model</span> <span class="o">=</span> <span class="n">lg</span>  <span class="c1"># All models need to be available in /path/to/TandemRepeats/tandemrepeats/data/paml/</span>
    <span class="p">[[</span><span class="n">indel</span><span class="p">]]</span>
        <span class="n">indelRatePerSite</span> <span class="o">=</span> <span class="mf">0.01</span>  <span class="c1"># What magnitude is the indel rate compared to the substitution rate?</span>
        <span class="n">ignore_gaps</span> <span class="o">=</span> <span class="kc">True</span>  <span class="c1"># Shall trailing gaps be ignored / not penalised?</span>
        <span class="n">gaps</span> <span class="o">=</span> <span class="n">row_wise</span>
        <span class="n">zipf</span> <span class="o">=</span> <span class="mf">1.821</span>
    <span class="p">[[</span><span class="n">optimisation</span><span class="p">]]</span>
        <span class="n">start_min</span> <span class="o">=</span> <span class="mf">0.5</span>
        <span class="n">start_max</span> <span class="o">=</span> <span class="mf">1.5</span>
        <span class="n">nIteration</span> <span class="o">=</span> <span class="mi">14</span>
    <span class="p">[[</span><span class="n">K80</span><span class="p">]]</span>
        <span class="n">kappa</span> <span class="o">=</span> <span class="mf">2.59</span>
    <span class="p">[[</span><span class="n">TN93</span><span class="p">]]</span>
        <span class="n">alpha_1</span> <span class="o">=</span> <span class="mf">0.3</span>
        <span class="n">alpha_2</span> <span class="o">=</span> <span class="mf">0.4</span>
        <span class="n">beta</span> <span class="o">=</span> <span class="mf">0.7</span>
<span class="p">[[</span><span class="n">score_calibration</span><span class="p">]]</span>
    <span class="n">scoreslist</span><span class="o">=</span><span class="n">phylo_gap01</span><span class="p">,</span>
    <span class="n">save_calibration</span> <span class="o">=</span> <span class="kc">False</span>
    <span class="n">precision</span> <span class="o">=</span> <span class="mi">10</span>
</pre></div>
</div>
</section>
<section id="creation-of-new-repeat-list-instances">
<h3>Creation of new Repeat list instances<a class="headerlink" href="#creation-of-new-repeat-list-instances" title="Permalink to this heading">¶</a></h3>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">repeat_list</span><span class="p">]</span>
    <span class="n">output_characteristics</span> <span class="o">=</span> <span class="n">begin</span><span class="p">,</span> <span class="n">msa_original</span><span class="p">,</span> <span class="n">l_effective</span><span class="p">,</span> <span class="n">n_effective</span><span class="p">,</span> <span class="n">repeat_region_length</span><span class="p">,</span> <span class="n">divergence</span><span class="p">,</span> <span class="n">pvalue</span>
    <span class="n">model</span> <span class="o">=</span> <span class="n">phylo_gap01</span>
</pre></div>
</div>
</section>
<section id="restrict-hmmer-model-size">
<h3>Restrict Hmmer model size.<a class="headerlink" href="#restrict-hmmer-model-size" title="Permalink to this heading">¶</a></h3>
<p>Set the maximum size of HMM for which the Viterbi algorithm is performed <em>l_effective_max</em> e.g.
to ensure viable run-times on your system:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">hmm</span><span class="p">]</span>
    <span class="n">l_effective_max</span> <span class="o">=</span> <span class="mi">50</span>
</pre></div>
</div>
</section>
</section>
<section id="logging-ini-optional">
<h2>logging.ini - OPTIONAL<a class="headerlink" href="#logging-ini-optional" title="Permalink to this heading">¶</a></h2>
<p>In this file, you can define the level of debugging per module (DEBUG, INFO, WARNING), and
the format of the debugging message. Defaults to WARNING. The path to the file needs to be
defined as</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">logging</span>
<span class="kn">import</span> <span class="nn">logging.config</span>
<span class="n">logging</span><span class="o">.</span><span class="n">config</span><span class="o">.</span><span class="n">fileConfig</span><span class="p">(</span><span class="s2">&quot;path/to/your/home/.tral/logging.ini&quot;</span><span class="p">)</span>
</pre></div>
</div>
</section>
</section>


            <div class="clearer"></div>
          </div>
        </div>
      </div>
      <div class="clearer"></div>
    </div>


  </body>
</html>