<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>Searching for particular repeats &mdash; TRAL 0.3.5 documentation</title>
    
    <link rel="stylesheet" href="_static/tral.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    './',
        VERSION:     '0.3.5',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <link rel="top" title="TRAL 0.3.5 documentation" href="index.html" />
    <link rel="prev" title="Use GC3Pie to annotate your large sequence dataset." href="workflow.html" />
    <link href='http://fonts.googleapis.com/css?family=Source+Sans+Pro:300,400,700|Source+Code+Pro|Armata|IM+Fell+English'
          rel='stylesheet' type='text/css' />
 
    <style type="text/css">
      table.right { float: right; margin-left: 20px; }
      table.right td { border: 1px solid #ccc; }
      
    </style>
    <script type="text/javascript">
      // intelligent scrolling of the sidebar content
      $(window).scroll(function() {
        var sb = $('.sphinxsidebarwrapper');
        var win = $(window);
        var sbh = sb.height();
        var offset = $('.sphinxsidebar').position()['top'];
        var wintop = win.scrollTop();
        var winbot = wintop + win.innerHeight();
        var curtop = sb.position()['top'];
        var curbot = curtop + sbh;
        // does sidebar fit in window?
        if (sbh < win.innerHeight()) {
          // yes: easy case -- always keep at the top
          sb.css('top', $u.min([$u.max([0, wintop - offset - 10]),
                                $(document).height() - sbh - 200]));
        } else {
          // no: only scroll if top/bottom edge of sidebar is at
          // top/bottom edge of window
          if (curtop > wintop && curbot > winbot) {
            sb.css('top', $u.max([wintop - offset - 10, 0]));
          } else if (curtop < wintop && curbot < winbot) {
            sb.css('top', $u.min([winbot - sbh - offset - 20,
                                  $(document).height() - sbh - 200]));
          }
        }
      });
    </script>

  </head>
  <body role="document">
<div class="pageheader">
    <a href="http://www.isb-sib.ch/" class="sib_logo" title="SIB Swiss Institute of Bioinformatics"><img src="_static/sib.png" alt="SIB" class="sib_logo" /></a>
    <a href="index.html">
      <img class="headlogo" src="_static/trallogo.png" alt="TRAL" />
      <span class="headlogo">Tandem Repeat Annotation Library</span>
    </a>
    <div class="headlinks">
        <a href="index.html">Home</a>
        <a href="install.html">Install</a>
        <a href="code_docs.html">Code docs</a>
    </div>
</div>


      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
  <h3><a href="index.html">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">Searching for particular repeats</a><ul>
<li><a class="reference internal" href="#overview">Overview</a></li>
<li><a class="reference internal" href="#id1">search_hmm</a><ul>
<li><a class="reference internal" href="#full-usage">Full usage</a></li>
</ul>
</li>
<li><a class="reference internal" href="#filter-hmm">filter_hmm</a><ul>
<li><a class="reference internal" href="#id2">Full usage</a></li>
</ul>
</li>
<li><a class="reference internal" href="#python-api">Python API</a></li>
</ul>
</li>
</ul>

  <h4>Previous topic</h4>
  <p class="topless"><a href="workflow.html"
                        title="previous chapter">Use GC3Pie to annotate your large sequence dataset.</a></p>
  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="_sources/search_hmm.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="section" id="searching-for-particular-repeats">
<span id="search-hmm"></span><h1>Searching for particular repeats<a class="headerlink" href="#searching-for-particular-repeats" title="Permalink to this headline">¶</a></h1>
<div class="section" id="overview">
<h2>Overview<a class="headerlink" href="#overview" title="Permalink to this headline">¶</a></h2>
<p>The search package provides python modules and command line tools for identifying
tandem repeats in sequence databases. Typically processing consists of first
running <code class="docutils literal"><span class="pre">search_hmm</span></code> to identify instances of a query repeat
in a sequence database. This is relatively slow (hours to days
against Uniprot) but can be trivially parallelized to run on a cluster.</p>
<p>After scoring the hmm against every sequence in the database, <code class="docutils literal"><span class="pre">filter_hmm</span></code> is
used to identify significant results.</p>
</div>
<div class="section" id="id1">
<h2>search_hmm<a class="headerlink" href="#id1" title="Permalink to this headline">¶</a></h2>
<p>The search module can be run directly:</p>
<div class="highlight-python"><div class="highlight"><pre>python -m tral.search.search_hmm -h
</pre></div>
</div>
<p>The query HMM should be in HMMR format. The <code class="docutils literal"><span class="pre">hmmbuild</span></code> tool may be helpful for
generating HMM files from multiple alignments. The HMM should represent a single
repeat. Circularly permuted edges are automatically inserted to permit multiple
sequential repeats.</p>
<p>The database should be in FASTA format. It may be gzip compressed.</p>
<p>Results are produced in TSV format, and include</p>
<ul class="simple">
<li>Sequence identifier (header line from FASTA file)</li>
<li>Viterbi probability (probability that this sequence was emitted by the query
HMM)</li>
<li>Log-odds ratio (viterbi probability divided by the expected probability for
a sequence of this length)</li>
<li>HMM states (Maximum likelihood sequence of states through the HMM, as
determined by the Viterbi algorithm)</li>
</ul>
<p>Example:</p>
<div class="highlight-python"><div class="highlight"><pre>python -m tral.search.search_hmm query.hmm database.faa.gz results.tsv
</pre></div>
</div>
<p>When running on a cluster, it may be useful to search just a subset of the
database. For instance, here&#8217;s a simplified submission script using the <code class="docutils literal"><span class="pre">-n</span></code>
and <code class="docutils literal"><span class="pre">-s</span></code> parameters in an array job:</p>
<div class="highlight-python"><div class="highlight"><pre>#!/bin/bash
#BSUB -J tral[1-10]
n=1000
start=$(( (${LSB_JOBINDEX:-1}-1) * $n ))
python -m tral.search.tral_search \
    -n $n -s $start \
    query.hmm database.fasta.gz results.${LSB_JOBINDEX:-1}.tsv
</pre></div>
</div>
<p>Choosing an appropriate log-odds ratio depends on the database size, amino
acid content, HMM length, and HMM composition. To help calibrate this, the <code class="docutils literal"><span class="pre">-r</span></code>
option is available to automatically shuffle the input sequences. The results
then represent all negative hits, allowing the False Discovery Rate to be
estimated for various thresholds in the filtering step.</p>
<div class="section" id="full-usage">
<h3>Full usage<a class="headerlink" href="#full-usage" title="Permalink to this headline">¶</a></h3>
<div class="highlight-python"><div class="highlight"><pre>usage: search_hmm.py [-h] [-v] [-s START] [-n DBSIZE] [-r]
                    hmm database results

Align a cpHMM against a database using TRAL

positional arguments:
hmm                   cpHMM filename
database              database to search, in fasta format (may be gzip
                        compressed)
results               results file

optional arguments:
-h, --help            show this help message and exit
-v, --verbose         Long messages
-s START, --start START
                        Index of database sequence to start with (default: 0)
-n DBSIZE, --dbsize DBSIZE
                        Number of database sequences to include (default: 0 to
                        include all)
-r, --shuffle         Shuffle database sequences
</pre></div>
</div>
</div>
</div>
<div class="section" id="filter-hmm">
<h2>filter_hmm<a class="headerlink" href="#filter-hmm" title="Permalink to this headline">¶</a></h2>
<p>The results from <code class="docutils literal"><span class="pre">search_hmm</span></code> include both significant and insignificant hits.
Use <code class="docutils literal"><span class="pre">filter_hmm</span></code> to filter the results based on length and log-odds criteria.</p>
<div class="highlight-python"><div class="highlight"><pre>python -m tral.search.filter_hmm -h
</pre></div>
</div>
<p><strong>Output</strong>. The tool can produce three kinds of output:</p>
<ul class="simple">
<li><code class="docutils literal"><span class="pre">-f</span></code> <strong>FASTA</strong>: Subset of the input database corresponding to significant hits</li>
<li><code class="docutils literal"><span class="pre">-o</span></code> <strong>TSV</strong>: Subset of the <code class="docutils literal"><span class="pre">search_hmm</span></code> results corresponding to
significant hits</li>
<li><code class="docutils literal"><span class="pre">-x</span></code> <strong>TREKS</strong>: All significant repeats in T-REKS format.</li>
</ul>
<p>The TREKS format is most useful for exporting results to other programs.</p>
<p><strong>Identifiers</strong>. By default, the identifier is extracted from the database
header lines (e.g. <code class="docutils literal"><span class="pre">&gt;tr|id|description...</span></code>). For compatibility with other
programs, the <code class="docutils literal"><span class="pre">--preserve-header</span></code> option can also be used to match sequences
based on the full header.</p>
<p><strong>Thresholds</strong>. Thresholds can be set for either:</p>
<ul class="simple">
<li><code class="docutils literal"><span class="pre">-r</span></code> <strong>Number of repeats</strong> (float). Repeats are computed based on the HMM match
states. This means that partial repeats are counted as a fraction of the
query length.</li>
<li><code class="docutils literal"><span class="pre">-t</span></code> <strong>Log-odds threshold</strong> (float). Results with log-odds less than this value
are filtered out.</li>
</ul>
<p><strong>Memory usage</strong>. Memory usage should be low if a single output is used. If
several output options are selected then the significant hits are stored in
memory. If this is problematic, <code class="docutils literal"><span class="pre">filter_hmm</span></code> can be run separately for
each desired output.</p>
<p><strong>Example:</strong></p>
<div class="highlight-python"><div class="highlight"><pre>python -m tral.search.filter_hmm \
    -f filtered.faa \
    -o filtered.tsv \
    -x filtered.tdr \
    -r 2 -t 8.0 \
    results.tsv database.faa.gz
</pre></div>
</div>
<div class="section" id="id2">
<h3>Full usage<a class="headerlink" href="#id2" title="Permalink to this headline">¶</a></h3>
<div class="highlight-python"><div class="highlight"><pre>usage: filter_hmm.py [-h] [-f FILTERED_FASTA] [-o FILTERED_TSV]
                    [-x FILTERED_TREKS] [-r MIN_REPEATS] [-t LOG_ODDS]
                    [--preserve-header] [-v]
                    hits database

Filter hits from search_hmm

positional arguments:
hits                  TSV file, as produced by search_hmm, containing HMM
                        hits
database              database to filter, in fasta format (may be gzip
                        compressed)

optional arguments:
-h, --help            show this help message and exit
--preserve-header     Include the full header in FASTA output. Otherwise,
                        just the identifier is used to match TSV and TREKS.
-v, --verbose         Long messages

outputs:
-f FILTERED_FASTA, --filtered-fasta FILTERED_FASTA
                        output fasta file, filtered by hits
-o FILTERED_TSV, --filtered-tsv FILTERED_TSV
                        Filtered TSV file
-x FILTERED_TREKS, --filtered-treks FILTERED_TREKS
                        Filtered T-Reks file

Thresholds:
-r MIN_REPEATS, --min-repeats MIN_REPEATS
                        Minimum number of repeats
-t LOG_ODDS, --log-odds LOG_ODDS
                        Threshold for minimum log-odds ratio
</pre></div>
</div>
</div>
</div>
<div class="section" id="python-api">
<h2>Python API<a class="headerlink" href="#python-api" title="Permalink to this headline">¶</a></h2>
<p>See also the <a class="reference internal" href="/tral/search.html"><em>Search code documentation</em></a></p>
</div>
</div>


          </div>
        </div>
      </div>
      <div class="clearer"></div>
    </div>


  </body>
</html>