<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>Perform statistical significance test of tandem repeats. &mdash; TRAL 0.3.5 documentation</title>
    
    <link rel="stylesheet" href="_static/tral.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    './',
        VERSION:     '0.3.5',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <link rel="top" title="TRAL 0.3.5 documentation" href="index.html" />
    <link rel="next" title="Perform overlap filtering of redundant tandem repeat annotations." href="overlap_filtering.html" />
    <link rel="prev" title="Annotate tandem repeats from sequence domain models." href="cphmm.html" />
    <link href='http://fonts.googleapis.com/css?family=Source+Sans+Pro:300,400,700|Source+Code+Pro|Armata|IM+Fell+English'
          rel='stylesheet' type='text/css' />
 
    <style type="text/css">
      table.right { float: right; margin-left: 20px; }
      table.right td { border: 1px solid #ccc; }
      
    </style>
    <script type="text/javascript">
      // intelligent scrolling of the sidebar content
      $(window).scroll(function() {
        var sb = $('.sphinxsidebarwrapper');
        var win = $(window);
        var sbh = sb.height();
        var offset = $('.sphinxsidebar').position()['top'];
        var wintop = win.scrollTop();
        var winbot = wintop + win.innerHeight();
        var curtop = sb.position()['top'];
        var curbot = curtop + sbh;
        // does sidebar fit in window?
        if (sbh < win.innerHeight()) {
          // yes: easy case -- always keep at the top
          sb.css('top', $u.min([$u.max([0, wintop - offset - 10]),
                                $(document).height() - sbh - 200]));
        } else {
          // no: only scroll if top/bottom edge of sidebar is at
          // top/bottom edge of window
          if (curtop > wintop && curbot > winbot) {
            sb.css('top', $u.max([wintop - offset - 10, 0]));
          } else if (curtop < wintop && curbot < winbot) {
            sb.css('top', $u.min([winbot - sbh - offset - 20,
                                  $(document).height() - sbh - 200]));
          }
        }
      });
    </script>

  </head>
  <body role="document">
<div class="pageheader">
    <a href="http://www.isb-sib.ch/" class="sib_logo" title="SIB Swiss Institute of Bioinformatics"><img src="_static/sib.png" alt="SIB" class="sib_logo" /></a>
    <a href="index.html">
      <img class="headlogo" src="_static/trallogo.png" alt="TRAL" />
      <span class="headlogo">Tandem Repeat Annotation Library</span>
    </a>
    <div class="headlinks">
        <a href="index.html">Home</a>
        <a href="install.html">Install</a>
        <a href="code_docs.html">Code docs</a>
    </div>
</div>


      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
  <h3><a href="index.html">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">Perform statistical significance test of tandem repeats.</a><ul>
<li><a class="reference internal" href="#read-in-tandem-repeat-annotations">Read in tandem repeat annotations.</a></li>
<li><a class="reference internal" href="#perform-a-statistical-significance-test">Perform a statistical significance test.</a></li>
<li><a class="reference internal" href="#filter-tandem-repeat-below-a-significance-threshold">Filter tandem repeat below a significance threshold.</a></li>
<li><a class="reference internal" href="#available-significance-tests">Available significance tests</a></li>
</ul>
</li>
</ul>

  <h4>Previous topic</h4>
  <p class="topless"><a href="cphmm.html"
                        title="previous chapter">Annotate tandem repeats from sequence domain models.</a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="overlap_filtering.html"
                        title="next chapter">Perform overlap filtering of redundant tandem repeat annotations.</a></p>
  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="_sources/significance_test.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="section" id="perform-statistical-significance-test-of-tandem-repeats">
<span id="significance-test"></span><h1>Perform statistical significance test of tandem repeats.<a class="headerlink" href="#perform-statistical-significance-test-of-tandem-repeats" title="Permalink to this headline">¶</a></h1>
<p>Here you learn how to perform statistical significance tests on your candidate set of tandem repeats.
TRAL currently provides a variety of model-based and none-model based tests. They are explained in 
detail in this <a class="reference internal" href="contributors.html#publications"><span>NAR publication (2012)</span></a>.</p>
<p>Requirements for this tutorial:</p>
<ul class="simple">
<li><a class="reference internal" href="install.html#install"><span>Install TRAL</span></a>. TRAL ships with the sequence data needed for this tutorial.</li>
<li><a class="reference internal" href="pvaluefiles.html#pvaluefiles"><span>Download p-value distribution files</span></a>.</li>
</ul>
<div class="section" id="read-in-tandem-repeat-annotations">
<h2>Read in tandem repeat annotations.<a class="headerlink" href="#read-in-tandem-repeat-annotations" title="Permalink to this headline">¶</a></h2>
<div class="highlight-python"><div class="highlight"><pre><span class="kn">import</span> <span class="nn">os</span><span class="o">,</span> <span class="nn">pickle</span>
<span class="kn">from</span> <span class="nn">tral</span> <span class="kn">import</span> <span class="n">sequence</span>
<span class="kn">from</span> <span class="nn">tral.paths</span> <span class="kn">import</span> <span class="n">PACKAGE_DIRECTORY</span>

<span class="n">fRepeat_Pickle</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">PACKAGE_DIRECTORY</span><span class="p">,</span><span class="s2">&quot;examples&quot;</span><span class="p">,</span> <span class="s2">&quot;data&quot;</span><span class="p">,</span> <span class="s2">&quot;HIV-1_388796.pickle&quot;</span><span class="p">)</span>

<span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">fRepeat_Pickle</span><span class="p">,</span> <span class="s1">&#39;rb&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">fh</span><span class="p">:</span>
    <span class="n">lHIV_Sequence</span> <span class="o">=</span> <span class="n">pickle</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="n">fh</span><span class="p">)</span>
</pre></div>
</div>
<p>There are nine proteins in the HIV proteome:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="nb">len</span><span class="p">(</span><span class="n">lHIV_Sequence</span><span class="p">)</span>
<span class="go">9</span>
</pre></div>
</div>
</div>
<div class="section" id="perform-a-statistical-significance-test">
<h2>Perform a statistical significance test.<a class="headerlink" href="#perform-a-statistical-significance-test" title="Permalink to this headline">¶</a></h2>
<p>Calculate the statistical score on the &#8216;denovo&#8217; <em>repeat_list</em>, and the corresponding
p-value with parameters defined in the <a class="reference internal" href="configure.html#configure"><span>configuration file</span></a>:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="k">for</span> <span class="n">iSequence</span> <span class="ow">in</span> <span class="n">lHIV_Sequence</span><span class="p">:</span>
    <span class="k">for</span> <span class="n">iTandemRepeat</span> <span class="ow">in</span> <span class="n">iSequence</span><span class="o">.</span><span class="n">get_repeatlist</span><span class="p">(</span><span class="s1">&#39;denovo&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">repeats</span><span class="p">:</span>
        <span class="n">iTandemRepeat</span><span class="o">.</span><span class="n">calculate_pvalues</span><span class="p">()</span>
</pre></div>
</div>
<p>For example, the following putative tandem repeat is found to be non-significant with the used model
of tandem repeat evolution <a class="reference internal" href="background.html#background"><span>(interpretation)</span></a>:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="k">print</span><span class="p">(</span><span class="n">lHIV_Sequence</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span><span class="o">.</span><span class="n">get_repeatlist</span><span class="p">(</span><span class="s1">&#39;denovo&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">repeats</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
<span class="go">&gt; begin:94 l_effective:7 n:2 pvalue:1.0 divergence:1.656005859375 type:phylo_gap01</span>
<span class="go">EKGGLEGLIYSKKRQE</span>
<span class="go">---ILDLWVY------</span>
</pre></div>
</div>
<p>Whereas this putative tandem repeat is considered significant:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="k">print</span><span class="p">(</span><span class="n">lHIV_Sequence</span><span class="p">[</span><span class="mi">4</span><span class="p">]</span><span class="o">.</span><span class="n">get_repeatlist</span><span class="p">(</span><span class="s1">&#39;denovo&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">repeats</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
<span class="go">&gt; begin:38 l_effective:2 n:6 pvalue:0.0 divergence:0.46649169922545164 type:phylo_gap01</span>
<span class="go">RRN</span>
<span class="go">RR-</span>
<span class="go">RRW</span>
<span class="go">RA-</span>
<span class="go">RQ-</span>
<span class="go">RQI</span>
</pre></div>
</div>
<p>Besides, parameters can also be directly supplied to <em>calculate_pvalues</em>:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="k">for</span> <span class="n">iSequence</span> <span class="ow">in</span> <span class="n">lHIV_Sequence</span><span class="p">:</span>
    <span class="k">for</span> <span class="n">iTandemRepeat</span> <span class="ow">in</span> <span class="n">iSequence</span><span class="o">.</span><span class="n">get_repeatlist</span><span class="p">(</span><span class="s1">&#39;denovo&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">repeats</span><span class="p">:</span>
        <span class="n">iTandemRepeat</span><span class="o">.</span><span class="n">calculate_pvalues</span><span class="p">(</span><span class="n">scoreslist</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;phylo_gap001&#39;</span><span class="p">])</span>
</pre></div>
</div>
</div>
<div class="section" id="filter-tandem-repeat-below-a-significance-threshold">
<h2>Filter tandem repeat below a significance threshold.<a class="headerlink" href="#filter-tandem-repeat-below-a-significance-threshold" title="Permalink to this headline">¶</a></h2>
<div class="highlight-python"><div class="highlight"><pre><span class="k">for</span> <span class="n">iSequence</span> <span class="ow">in</span> <span class="n">lHIV_Sequence</span><span class="p">:</span>
    <span class="n">repeat_list</span> <span class="o">=</span> <span class="n">iSequence</span><span class="o">.</span><span class="n">get_repeatlist</span><span class="p">(</span><span class="s1">&#39;denovo&#39;</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">repeat_list</span><span class="p">:</span>
        <span class="n">repeat_list_filtered</span> <span class="o">=</span> <span class="n">repeat_list</span><span class="o">.</span><span class="n">filter</span><span class="p">(</span><span class="n">func_name</span> <span class="o">=</span> <span class="s2">&quot;pvalue&quot;</span><span class="p">,</span> <span class="n">score</span> <span class="o">=</span> <span class="s2">&quot;phylo_gap01&quot;</span><span class="p">,</span> <span class="n">threshold</span> <span class="o">=</span> <span class="mf">0.05</span><span class="p">)</span>
        <span class="n">iSequence</span><span class="o">.</span><span class="n">set_repeatlist</span><span class="p">(</span><span class="n">repeat_list_filtered</span><span class="p">,</span> <span class="s2">&quot;denovo_filtered&quot;</span><span class="p">)</span>
</pre></div>
</div>
<p>The resulting <em>result_list</em> now only contains tandem repeats with a p-value below
0.05, and is shorter than the original tandem repeat list:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="nb">len</span><span class="p">([</span><span class="n">iR</span> <span class="k">for</span> <span class="n">iS</span> <span class="ow">in</span> <span class="n">lHIV_Sequence</span> <span class="k">for</span> <span class="n">iR</span> <span class="ow">in</span> <span class="n">iS</span><span class="o">.</span><span class="n">get_repeatlist</span><span class="p">(</span><span class="s1">&#39;denovo&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">repeats</span><span class="p">])</span>
<span class="go">28</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">len</span><span class="p">([</span><span class="n">iR</span> <span class="k">for</span> <span class="n">iS</span> <span class="ow">in</span> <span class="n">lHIV_Sequence</span> <span class="k">for</span> <span class="n">iR</span> <span class="ow">in</span> <span class="n">iS</span><span class="o">.</span><span class="n">get_repeatlist</span><span class="p">(</span><span class="s1">&#39;denovo_filtered&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">repeats</span><span class="p">])</span>
<span class="go">20</span>
</pre></div>
</div>
</div>
<div class="section" id="available-significance-tests">
<h2>Available significance tests<a class="headerlink" href="#available-significance-tests" title="Permalink to this headline">¶</a></h2>
<p>The model based scores are generally described in our <a class="reference internal" href="contributors.html#publications"><span>NAR publication, 2012</span></a>.
Briefly, the sequences evolution is assumed to be described by the LG matrix.</p>
<p>Further, the models can be distinguished by how gaps are handled:</p>
<ul class="simple">
<li>&#8216;phylo&#8217;: Gaps are treated as characters.</li>
</ul>
<p>Gaps can also assumed to be  inserted with exponentially distributed waiting times and Zipfian distributed indel lengths:</p>
<ul class="simple">
<li>&#8216;phylo_gap01&#8217;, &#8216;phylo_gap01_ignore_trailing_gaps&#8217;, &#8216;phylo_gap01_ignore_coherent_deletions&#8217;,
&#8216;phylo_gap01_ignore_trailing_gaps_and_coherent_deletions&#8217;: Gaps are modelled, and assumed to mutation at at 0.1 times lower rate compared to substitutions.</li>
<li>&#8216;phylo_gap001&#8217;, &#8216;phylo_gap001_ignore_trailing_gaps&#8217;, &#8216;phylo_gap001_ignore_coherent_deletions&#8217;,
&#8216;phylo_gap001_ignore_trailing_gaps_and_coherent_deletions&#8217;: Gaps are modelled, and assumed to mutation  at at 0.01 times lower rate compared to substitutions.</li>
<li><em>ignore_trailing_gaps</em> signifies that gaps before the first tandem repeat unit and after the last tandem repeat unit are not penalised.</li>
<li><em>coherent_deletions</em> signifies that gaps of the same length and position within the alignment of tandem repeat units are penalised only once.</li>
</ul>
<p>Among the three <em>ad hoc</em> scores are &#8216;entropy&#8217;, &#8216;parsimony&#8217;, and &#8216;pSim&#8217; scores
(see our <a class="reference internal" href="contributors.html#publications"><span>NAR publication, 2012</span></a>).</p>
</div>
</div>


          </div>
        </div>
      </div>
      <div class="clearer"></div>
    </div>


  </body>
</html>