
<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>tral.repeat.repeat_io &#8212; TRAL 2.0.0 documentation</title>
    <link rel="stylesheet" href="../../../_static/tral.css" type="text/css" />
    <link rel="stylesheet" href="../../../_static/pygments.css" type="text/css" />
    <script id="documentation_options" data-url_root="../../../" src="../../../_static/documentation_options.js"></script>
    <script src="../../../_static/jquery.js"></script>
    <script src="../../../_static/underscore.js"></script>
    <script src="../../../_static/doctools.js"></script>
    <script src="../../../_static/language_data.js"></script>
    <link rel="index" title="Index" href="../../../genindex.html" />
    <link rel="search" title="Search" href="../../../search.html" />
    <link href='http://fonts.googleapis.com/css?family=Source+Sans+Pro:300,400,700|Source+Code+Pro|Armata|IM+Fell+English'
          rel='stylesheet' type='text/css' />
 
    <style type="text/css">
      table.right { float: right; margin-left: 20px; }
      table.right td { border: 1px solid #ccc; }
      
    </style>
    <script type="text/javascript">
      // intelligent scrolling of the sidebar content
      $(window).scroll(function() {
        var sb = $('.sphinxsidebarwrapper');
        var win = $(window);
        var sbh = sb.height();
        var offset = $('.sphinxsidebar').position()['top'];
        var wintop = win.scrollTop();
        var winbot = wintop + win.innerHeight();
        var curtop = sb.position()['top'];
        var curbot = curtop + sbh;
        // does sidebar fit in window?
        if (sbh < win.innerHeight()) {
          // yes: easy case -- always keep at the top
          sb.css('top', $u.min([$u.max([0, wintop - offset - 10]),
                                $(document).height() - sbh - 200]));
        } else {
          // no: only scroll if top/bottom edge of sidebar is at
          // top/bottom edge of window
          if (curtop > wintop && curbot > winbot) {
            sb.css('top', $u.max([wintop - offset - 10, 0]));
          } else if (curtop < wintop && curbot < winbot) {
            sb.css('top', $u.min([winbot - sbh - offset - 20,
                                  $(document).height() - sbh - 200]));
          }
        }
      });
    </script>

  </head><body>
<div class="pageheader">
    <a href="http://www.isb-sib.ch/" class="sib_logo" title="SIB Swiss Institute of Bioinformatics"><img src="../../../_static/sib.png" alt="SIB" class="sib_logo" /></a>
    <a href="../../../index.html">
      <img class="headlogo" src="../../../_static/trallogo.png" alt="TRAL" />
      <span class="headlogo">Tandem Repeat Annotation Library</span>
    </a>
    <div class="headlinks">
        <a href="../../../index.html">Home</a>
        <a href="../../../install.html">Install</a>
        <a href="../../../code_docs.html">Code docs</a>
    </div>
</div>


      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
<div id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="../../../search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" />
      <input type="submit" value="Go" />
    </form>
    </div>
</div>
<script>$('#searchbox').show(0);</script>
        </div>
      </div>

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <h1>Source code for tral.repeat.repeat_io</h1><div class="highlight"><pre>
<span></span><span class="c1"># (C) 2015 Elke Schaper</span>

<span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    :synopsis: Input/output for tandem repeats.</span>

<span class="sd">    .. moduleauthor:: Elke Schaper &lt;elke.schaper@isb-sib.ch&gt;</span>
<span class="sd">&quot;&quot;&quot;</span>

<span class="kn">import</span> <span class="nn">Bio.Seq</span>
<span class="kn">import</span> <span class="nn">random</span>
<span class="kn">import</span> <span class="nn">logging</span>
<span class="kn">import</span> <span class="nn">itertools</span>
<span class="kn">import</span> <span class="nn">os</span>
<span class="kn">import</span> <span class="nn">re</span>
<span class="kn">import</span> <span class="nn">shutil</span>
<span class="kn">import</span> <span class="nn">tempfile</span>
<span class="kn">import</span> <span class="nn">subprocess</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="kn">from</span> <span class="nn">tral.repeat</span> <span class="kn">import</span> <span class="n">repeat</span>
<span class="kn">from</span> <span class="nn">tral</span> <span class="kn">import</span> <span class="n">configuration</span>

<span class="kn">from</span> <span class="nn">tral.paths</span> <span class="kn">import</span> <span class="n">config_file</span>
<span class="n">CONFIG_GENERAL</span> <span class="o">=</span> <span class="n">configuration</span><span class="o">.</span><span class="n">Configuration</span><span class="o">.</span><span class="n">instance</span><span class="p">()</span><span class="o">.</span><span class="n">config</span>
<span class="n">REPEAT_CONFIG</span> <span class="o">=</span> <span class="n">CONFIG_GENERAL</span><span class="p">[</span><span class="s2">&quot;repeat&quot;</span><span class="p">]</span>

<span class="n">LOG</span> <span class="o">=</span> <span class="n">logging</span><span class="o">.</span><span class="n">getLogger</span><span class="p">(</span><span class="vm">__name__</span><span class="p">)</span>


<span class="c1">###################### SAVE REPEAT #######################################</span>

<div class="viewcode-block" id="save_repeat_fasta"><a class="viewcode-back" href="../../../tral/repeat.html#tral.repeat.repeat_io.save_repeat_fasta">[docs]</a><span class="k">def</span> <span class="nf">save_repeat_fasta</span><span class="p">(</span><span class="n">tandem_repeats</span><span class="p">,</span> <span class="n">file</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot; save multiple &lt;tandem_repeats&gt; in Fasta format in specified &lt;file&gt;</span>

<span class="sd">        At current, only one TR per sequence can be defined, as the identifiers</span>
<span class="sd">        in the dict &lt;tandem_repeats&gt; must be unique.</span>

<span class="sd">        Parameters: Dict of tandem repeats and identifiers.</span>
<span class="sd">            e.g. {&#39;ENSP00012&#39;: msa1, &#39;ENSP00013&#39;: msa2}</span>

<span class="sd">            &gt;ID</span>
<span class="sd">            GHKI</span>
<span class="sd">            GHKI</span>
<span class="sd">            GH--</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">file</span><span class="p">,</span> <span class="s1">&#39;w&#39;</span><span class="p">,</span> <span class="n">newline</span><span class="o">=</span><span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">identifier</span><span class="p">,</span> <span class="n">msa</span> <span class="ow">in</span> <span class="n">tandem_repeats</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
            <span class="n">f</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s2">&quot;&gt;</span><span class="si">{0}</span><span class="se">\n</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">identifier</span><span class="p">))</span>
            <span class="n">f</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">msa</span><span class="p">)</span> <span class="o">+</span> <span class="s2">&quot;</span><span class="se">\n\n</span><span class="s2">&quot;</span><span class="p">)</span></div>


<div class="viewcode-block" id="save_repeat_stockholm"><a class="viewcode-back" href="../../../tral/repeat.html#tral.repeat.repeat_io.save_repeat_stockholm">[docs]</a><span class="k">def</span> <span class="nf">save_repeat_stockholm</span><span class="p">(</span><span class="n">tandem_repeat</span><span class="p">,</span> <span class="n">file</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot; save &lt;tandem_repeat&gt; in STOCKHOLM format in specified &lt;file&gt;</span>

<span class="sd">        Parameters: Tandem repeat MSA.</span>
<span class="sd">            e.g. [&quot;ACDEF-&quot;, &quot;ACCDEF&quot;]</span>

<span class="sd">            More information in</span>
<span class="sd">            ftp://selab.janelia.org/pub/software/hmmer3/3.0/Userguide.pdf</span>
<span class="sd">            STOCKHOLM Format Example:</span>

<span class="sd">            # STOCKHOLM 1.0</span>
<span class="sd">            seq1 ACDEF...GHIKL</span>
<span class="sd">            seq2 ACDEF...GHIKL</span>
<span class="sd">            seq3 ...EFMNRGHIKL</span>

<span class="sd">            seq1 MNPQTVWY</span>
<span class="sd">            seq2 MNPQTVWY</span>
<span class="sd">            seq3 MNPQT...</span>
<span class="sd">            //</span>

<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">file</span><span class="p">,</span> <span class="s1">&#39;w&#39;</span><span class="p">,</span> <span class="n">newline</span><span class="o">=</span><span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">fh</span><span class="p">:</span>
        <span class="n">fh</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s2">&quot;# STOCKHOLM 1.0</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">iMSA</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">tandem_repeat</span><span class="p">):</span>
            <span class="n">fh</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">{}</span><span class="s2"> </span><span class="si">{}</span><span class="se">\n</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">i</span><span class="p">),</span> <span class="n">iMSA</span><span class="p">))</span>
        <span class="n">fh</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s2">&quot;//&quot;</span><span class="p">)</span></div>


<div class="viewcode-block" id="save_repeat_treks"><a class="viewcode-back" href="../../../tral/repeat.html#tral.repeat.repeat_io.save_repeat_treks">[docs]</a><span class="k">def</span> <span class="nf">save_repeat_treks</span><span class="p">(</span><span class="n">tandem_repeats</span><span class="p">,</span> <span class="n">file</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot; Save multiple `tandem_repeats` in T-REKS format in specified `file`</span>

<span class="sd">        At current, only one TR per sequence can be defined, as the identifiers</span>
<span class="sd">        in the dict &lt;tandem_repeats&gt; must be unique.</span>

<span class="sd">        Parameters: Dict of tandem repeats and identifiers.</span>
<span class="sd">            e.g. {&#39;ENSP00012&#39;: [msa1, begin1], &#39;ENSP00013&#39;: [msa2, begin2]}</span>

<span class="sd">            T-REKS format example:</span>

<span class="sd">            &gt;a</span>
<span class="sd">            Length: 3 residues - nb: 3  from  1 to 10 - Psim:0.8076923076923077 region Length:42</span>
<span class="sd">            GHKI</span>
<span class="sd">            GHKI</span>
<span class="sd">            GH--</span>
<span class="sd">            **********************</span>

<span class="sd">            Length: 3 residues - nb: 2  from  21 to 27 - Psim:0.7857142857142857 region Length:22</span>
<span class="sd">            GHKI</span>
<span class="sd">            GH--</span>
<span class="sd">            **********************</span>

<span class="sd">            &gt;b</span>
<span class="sd">            Length: 3 residues - nb: 3  from  1 to 10 - Psim:0.8095238095238095 region Length:20</span>
<span class="sd">            UIR</span>
<span class="sd">            UIR</span>
<span class="sd">            UIR</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">file</span><span class="p">,</span> <span class="s1">&#39;w&#39;</span><span class="p">,</span> <span class="n">newline</span><span class="o">=</span><span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">fh</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">identifier</span><span class="p">,</span> <span class="n">info</span> <span class="ow">in</span> <span class="n">tandem_repeats</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
            <span class="n">fh</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s2">&quot;&gt;</span><span class="si">{0}</span><span class="se">\n</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">identifier</span><span class="p">))</span>
            <span class="n">fh</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s2">&quot;Length: from </span><span class="si">{0}</span><span class="s2"> to</span><span class="se">\n</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">info</span><span class="p">[</span><span class="mi">1</span><span class="p">])))</span>
            <span class="n">fh</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">info</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span> <span class="o">+</span> <span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span> <span class="o">+</span> <span class="s2">&quot;*&quot;</span> <span class="o">*</span> <span class="mi">22</span> <span class="o">+</span> <span class="s2">&quot;</span><span class="se">\n\n</span><span class="s2">&quot;</span><span class="p">)</span></div>


<span class="c1">############################## READ REPEAT SEQUENCE ######################</span>


<div class="viewcode-block" id="read_fasta"><a class="viewcode-back" href="../../../tral/repeat.html#tral.repeat.repeat_io.read_fasta">[docs]</a><span class="k">def</span> <span class="nf">read_fasta</span><span class="p">(</span><span class="n">seq_filename</span><span class="p">,</span> <span class="n">sequence_type</span><span class="o">=</span><span class="s1">&#39;AA&#39;</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot; Read repeat from file in fasta format.</span>

<span class="sd">    Read repeat from file in fasta format.</span>

<span class="sd">    Args:</span>
<span class="sd">        seq_filename (str):  Path to the repeats containing fasta file</span>
<span class="sd">        sequence_type (str): Either &quot;AA&quot; or &quot;DNA&quot;</span>

<span class="sd">    Returns:</span>
<span class="sd">        (list of Repeat): A list of Repeat instances.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">pat_start</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">compile</span><span class="p">(</span><span class="sa">r</span><span class="s2">&quot;&gt;(.*)&quot;</span><span class="p">)</span>
    <span class="n">pat_repeat_unit</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">compile</span><span class="p">(</span><span class="sa">r</span><span class="s2">&quot;([\w\.\-]+)&quot;</span><span class="p">)</span>

    <span class="c1"># Our possible parser states:</span>
    <span class="c1">#</span>
    <span class="c1"># 1: searching for sequence name</span>
    <span class="c1"># 2: searching for repeat units</span>

    <span class="n">repeats</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="n">state</span> <span class="o">=</span> <span class="mi">1</span>
    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">seq_filename</span><span class="p">,</span> <span class="s2">&quot;rt&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">infile</span><span class="p">:</span>

        <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">line</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">infile</span><span class="p">):</span>
            <span class="n">LOG</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="s2">&quot;Line </span><span class="si">{0}</span><span class="s2">: </span><span class="si">{1}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">i</span><span class="p">,</span> <span class="n">line</span><span class="p">[</span><span class="mi">0</span><span class="p">:</span><span class="o">-</span><span class="mi">1</span><span class="p">]))</span>
            <span class="k">if</span> <span class="mi">1</span> <span class="o">==</span> <span class="n">state</span><span class="p">:</span>
                <span class="n">match</span> <span class="o">=</span> <span class="n">pat_start</span><span class="o">.</span><span class="n">match</span><span class="p">(</span><span class="n">line</span><span class="p">)</span>
                <span class="k">if</span> <span class="n">match</span><span class="p">:</span>
                    <span class="n">LOG</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="s2">&quot; * (1-&gt;2) Found start&quot;</span><span class="p">)</span>
                    <span class="n">LOG</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="s2">&quot;Start: </span><span class="si">%s</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">match</span><span class="o">.</span><span class="n">group</span><span class="p">(</span><span class="mi">1</span><span class="p">))</span>
                    <span class="n">name</span> <span class="o">=</span> <span class="n">match</span><span class="o">.</span><span class="n">group</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span>
                    <span class="n">repeats</span><span class="p">[</span><span class="n">name</span><span class="p">]</span> <span class="o">=</span> <span class="p">[]</span>
                    <span class="n">state</span> <span class="o">=</span> <span class="mi">2</span>

            <span class="k">elif</span> <span class="mi">2</span> <span class="o">==</span> <span class="n">state</span><span class="p">:</span>
                <span class="n">match</span> <span class="o">=</span> <span class="n">pat_repeat_unit</span><span class="o">.</span><span class="n">match</span><span class="p">(</span><span class="n">line</span><span class="p">)</span>
                <span class="k">if</span> <span class="n">match</span><span class="p">:</span>
                    <span class="n">LOG</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="s2">&quot; * (2-&gt;2) Found Repeat unit&quot;</span><span class="p">)</span>
                    <span class="n">LOG</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="s2">&quot;Repeat Unit: </span><span class="si">%s</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">match</span><span class="o">.</span><span class="n">group</span><span class="p">(</span><span class="mi">1</span><span class="p">))</span>
                    <span class="n">repeat_unit</span> <span class="o">=</span> <span class="n">match</span><span class="o">.</span><span class="n">group</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span>
                    <span class="n">repeats</span><span class="p">[</span><span class="n">name</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">repeat_unit</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot;.&quot;</span><span class="p">,</span> <span class="s2">&quot;-&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">upper</span><span class="p">())</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="n">LOG</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="s2">&quot; * (2-&gt;1) Found NO further Repeat unit&quot;</span><span class="p">)</span>
                    <span class="n">state</span> <span class="o">=</span> <span class="mi">1</span>

    <span class="k">for</span> <span class="n">i_name</span><span class="p">,</span> <span class="n">iR</span> <span class="ow">in</span> <span class="n">repeats</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
        <span class="k">yield</span> <span class="n">iR</span><span class="p">,</span> <span class="n">sequence_type</span></div>


<span class="c1">################################### SIMULATE SEQUENCE ####################</span>

<div class="viewcode-block" id="evolved_tandem_repeats"><a class="viewcode-back" href="../../../tral/repeat.html#tral.repeat.repeat_io.evolved_tandem_repeats">[docs]</a><span class="k">def</span> <span class="nf">evolved_tandem_repeats</span><span class="p">(</span><span class="n">l</span><span class="p">,</span> <span class="n">n</span><span class="p">,</span> <span class="n">n_samples</span><span class="p">,</span> <span class="n">sequence_type</span><span class="p">,</span> <span class="n">job_id</span><span class="o">=</span><span class="s1">&#39;job_id&#39;</span><span class="p">,</span>
                           <span class="n">mutation_rate</span><span class="o">=</span><span class="mi">50</span><span class="p">,</span> <span class="n">tree</span><span class="o">=</span><span class="s1">&#39;star&#39;</span><span class="p">,</span>
                           <span class="n">indel_rate_per_site</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
                           <span class="n">return_type</span><span class="o">=</span><span class="s1">&#39;repeat&#39;</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot; Simulate evolved sequences with ALF.</span>

<span class="sd">    Simulate evolved sequences with ALF:</span>
<span class="sd">    Dalquen, D. A., Anisimova, M., Gonnet, G. H. &amp; Dessimoz, C.</span>
<span class="sd">    ALF--a simulation framework for genome evolution. Molecular Biology and Evolution 29,</span>
<span class="sd">    1115–1123 (2012).</span>

<span class="sd">    Args:</span>
<span class="sd">        l (int): The length of the repeat unit</span>
<span class="sd">        n (int): The number of repeat units in the tandem repeat</span>
<span class="sd">        n_samples (int):  The number of samples</span>
<span class="sd">        sequence_type (str): Either &quot;AA&quot; or &quot;DNA&quot;</span>
<span class="sd">        job_id (str): A tag for files produces with ALF, and result files.</span>
<span class="sd">        mutation_rate (float): The mutation rate.</span>
<span class="sd">        tree (str): The type of tree, e.g. &quot;star&quot; or &quot;birthdeath&quot;</span>
<span class="sd">        indel_rate_per_site (int or False): The indel rate per site.</span>
<span class="sd">        return_type (str): Either &quot;repeat&quot; or &quot;list&quot;</span>

<span class="sd">        sequence_length (int): The total length of the simulated sequence</span>

<span class="sd">    Returns:</span>
<span class="sd">        Return type depends on ``return_type``: ``Repeat`` or ``Bio.Seq.Seq`` instance.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">runfile_template</span> <span class="o">=</span> <span class="n">config_file</span><span class="p">(</span><span class="s2">&quot;data&quot;</span><span class="p">,</span> <span class="s2">&quot;ALF&quot;</span><span class="p">,</span> <span class="s2">&quot;template.drw&quot;</span><span class="p">)</span>
    <span class="n">alf_exec</span> <span class="o">=</span> <span class="n">REPEAT_CONFIG</span><span class="p">[</span><span class="s1">&#39;alfsim&#39;</span><span class="p">]</span>
    <span class="c1"># create temporary directory</span>
    <span class="n">working_dir</span> <span class="o">=</span> <span class="n">tempfile</span><span class="o">.</span><span class="n">mkdtemp</span><span class="p">()</span>
    <span class="n">LOG</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="s2">&quot;evolvedTR: Created tempfile: </span><span class="si">%s</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">working_dir</span><span class="p">)</span>

    <span class="c1"># create working dir</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">isdir</span><span class="p">(</span><span class="n">working_dir</span><span class="p">):</span>
        <span class="n">os</span><span class="o">.</span><span class="n">makedirs</span><span class="p">(</span><span class="n">working_dir</span><span class="p">)</span>

    <span class="c1"># Copy template file and append job specific info</span>
    <span class="n">runfilename</span> <span class="o">=</span> <span class="s2">&quot;alf.drw&quot;</span>
    <span class="n">shutil</span><span class="o">.</span><span class="n">copyfile</span><span class="p">(</span><span class="n">runfile_template</span><span class="p">,</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">working_dir</span><span class="p">,</span> <span class="n">runfilename</span><span class="p">))</span>

    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">working_dir</span><span class="p">,</span> <span class="n">runfilename</span><span class="p">),</span> <span class="s2">&quot;a&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">runfile</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">indel_rate_per_site</span><span class="p">:</span>
            <span class="c1"># insertion rate per site and PAM. E.g. for PAM=40 expect</span>
            <span class="c1"># aaGainRate*40 insertions per site.</span>
            <span class="n">runfile</span><span class="o">.</span><span class="n">write</span><span class="p">(</span>
                <span class="s2">&quot;aaGainRate := &quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">indel_rate_per_site</span> <span class="o">/</span> <span class="n">mutation_rate</span><span class="p">)</span> <span class="o">+</span> <span class="s2">&quot;;</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="n">runfile</span><span class="o">.</span><span class="n">write</span><span class="p">(</span>
                <span class="s2">&quot;aaLossRate := &quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">indel_rate_per_site</span> <span class="o">/</span> <span class="n">mutation_rate</span><span class="p">)</span> <span class="o">+</span> <span class="s2">&quot;;</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="n">runfile</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s2">&quot;maxIndelLength := 50;</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="n">runfile</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s2">&quot;indelModel := &#39;ZIPF&#39;;</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="n">runfile</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s2">&quot;Z_c := 1.821;</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="n">runfile</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s2">&quot;DawgPlacement := true;</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="n">runfile</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s2">&quot;uuid := &#39;&quot;</span> <span class="o">+</span> <span class="n">job_id</span> <span class="o">+</span> <span class="s2">&quot;&#39;;</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="n">runfile</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s2">&quot;mname := &#39;&quot;</span> <span class="o">+</span> <span class="n">job_id</span> <span class="o">+</span> <span class="s2">&quot;&#39;;</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="n">runfile</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s2">&quot;protStart := &quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">n_samples</span><span class="p">)</span> <span class="o">+</span> <span class="s2">&quot;;</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="n">runfile</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s2">&quot;NSpecies := &quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">n</span><span class="p">)</span> <span class="o">+</span> <span class="s2">&quot;;</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="n">runfile</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s2">&quot;minGeneLength := &quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">l</span><span class="p">)</span> <span class="o">+</span> <span class="s2">&quot;;</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="n">runfile</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s2">&quot;mutRate := &quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">mutation_rate</span><span class="p">)</span> <span class="o">+</span> <span class="s2">&quot;;</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="n">runfile</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s2">&quot;wdir := &#39;&quot;</span> <span class="o">+</span> <span class="n">working_dir</span> <span class="o">+</span> <span class="s2">&quot;&#39;;</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>

        <span class="n">tree_length</span> <span class="o">=</span> <span class="n">n</span> <span class="o">*</span> <span class="n">mutation_rate</span>
        <span class="c1"># parameters concerning the species tree</span>
        <span class="k">if</span> <span class="n">tree</span> <span class="o">==</span> <span class="s1">&#39;star&#39;</span><span class="p">:</span>
            <span class="c1"># BDTree, ToLSample, Custom</span>
            <span class="n">runfile</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s2">&quot;treeType := &#39;Custom&#39;:</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="n">runfile</span><span class="o">.</span><span class="n">write</span><span class="p">(</span>
                <span class="s2">&quot;tree := MakeStarTree(BirthDeathTree(0.1,0.1,&quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">n</span><span class="p">)</span> <span class="o">+</span> <span class="s2">&quot;,10),mutRate):</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="n">runfile</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s2">&quot;treeLength:= </span><span class="si">%d</span><span class="s2"> ;</span><span class="se">\n</span><span class="s2">&quot;</span> <span class="o">%</span> <span class="n">tree_length</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="c1"># BDTree, ToLSample, Custom</span>
            <span class="n">runfile</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s2">&quot;treeType := &#39;BDTree&#39;:</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="c1"># From the last discussion with Daniel, only the ration of birth to</span>
            <span class="c1"># death rate matters, if we scale tree to match pam distance</span>
            <span class="n">runfile</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s2">&quot;birthRate := 0.01:</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="n">runfile</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s2">&quot;deathRate := 0.01:</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="c1"># for BDTree: should resulting tree be ultrametric, e.g. all leaves</span>
            <span class="c1"># have same distance to origin?</span>
            <span class="n">runfile</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s2">&quot;ultrametric := false:</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="n">runfile</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s2">&quot;treeLength:= </span><span class="si">%d</span><span class="s2"> ;</span><span class="se">\n</span><span class="s2">&quot;</span> <span class="o">%</span> <span class="n">tree_length</span><span class="p">)</span>
            <span class="c1"># DANIEL: Is a tree := missing?</span>
        <span class="k">if</span> <span class="n">tree</span> <span class="ow">not</span> <span class="ow">in</span> <span class="p">{</span><span class="s1">&#39;star&#39;</span><span class="p">,</span> <span class="s1">&#39;birthdeath&#39;</span><span class="p">}:</span>
            <span class="n">LOG</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span>
                <span class="s2">&quot;evolvedTR: tree input </span><span class="si">%s</span><span class="s2"> not known, assuming birthdeath tree&quot;</span><span class="p">,</span>
                <span class="n">tree</span><span class="p">)</span>

        <span class="c1"># parameters concerning the substitution models</span>
        <span class="k">if</span> <span class="n">sequence_type</span> <span class="o">==</span> <span class="s1">&#39;AA&#39;</span><span class="p">:</span>
            <span class="n">runfile</span><span class="o">.</span><span class="n">write</span><span class="p">(</span>
                <span class="s2">&quot;substModels := [SubstitutionModel(&#39;CustomP&#39;, [&#39;&quot;</span> <span class="o">+</span>
                <span class="n">config_file</span><span class="p">(</span><span class="s1">&#39;data&#39;</span><span class="p">,</span> <span class="s1">&#39;ALF&#39;</span><span class="p">,</span> <span class="s1">&#39;lg.dat&#39;</span><span class="p">)</span> <span class="o">+</span>
                <span class="s2">&quot;&#39;])];</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="c1"># CHECK! DOES</span>
            <span class="c1"># substModels := [SubstitutionModel(&#39;LG&#39;)];</span>
            <span class="c1"># WORK?</span>
        <span class="k">elif</span> <span class="n">sequence_type</span> <span class="o">==</span> <span class="s1">&#39;DNA&#39;</span><span class="p">:</span>
            <span class="n">runfile</span><span class="o">.</span><span class="n">write</span><span class="p">(</span>
                <span class="s2">&quot;substModels := [SubstitutionModel(&#39;TN93&#39;, [.3, .4, .7],[seq(0.25,4)], true)]:</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="c1"># CHECK! DO THE EMPIRICAL PARAMETERS MAKE SENSE AT ALL?</span>
            <span class="c1"># compare to http://people.inf.ethz.ch/ddalquen/alf/ALF_manual.pdf</span>
        <span class="k">else</span><span class="p">:</span>  <span class="c1"># CODONS</span>
            <span class="n">runfile</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s2">&quot;substModels := [SubstitutionModel(&#39;CPAM&#39;)];</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>

    <span class="c1"># Determine ALF MSA output file path</span>
    <span class="k">if</span> <span class="n">sequence_type</span> <span class="o">==</span> <span class="s1">&#39;AA&#39;</span><span class="p">:</span>
        <span class="n">infile</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">working_dir</span><span class="p">,</span> <span class="n">job_id</span><span class="p">,</span> <span class="s2">&quot;MSA&quot;</span><span class="p">,</span> <span class="s2">&quot;MSA_all_aa.phy&quot;</span><span class="p">)</span>
    <span class="k">elif</span> <span class="n">sequence_type</span> <span class="o">==</span> <span class="s1">&#39;DNA&#39;</span><span class="p">:</span>
        <span class="n">infile</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">working_dir</span><span class="p">,</span> <span class="n">job_id</span><span class="p">,</span> <span class="s2">&quot;MSA&quot;</span><span class="p">,</span> <span class="s2">&quot;MSA_all_dna.phy&quot;</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>  <span class="c1"># CODONS</span>
        <span class="n">infile</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span>
            <span class="n">working_dir</span><span class="p">,</span>
            <span class="n">job_id</span><span class="p">,</span>
            <span class="s2">&quot;MSA&quot;</span><span class="p">,</span>
            <span class="s2">&quot;MSA_all_codon.phy&quot;</span><span class="p">)</span>

    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">10</span><span class="p">):</span>
        <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">working_dir</span><span class="p">,</span> <span class="s2">&quot;out.txt&quot;</span><span class="p">),</span> <span class="s2">&quot;w&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">outfile</span><span class="p">:</span>
            <span class="n">alf_process</span> <span class="o">=</span> <span class="n">subprocess</span><span class="o">.</span><span class="n">Popen</span><span class="p">([</span><span class="n">alf_exec</span><span class="p">,</span>
                                            <span class="n">runfilename</span><span class="p">],</span>
                                           <span class="n">cwd</span><span class="o">=</span><span class="n">working_dir</span><span class="p">,</span>
                                           <span class="n">stdout</span><span class="o">=</span><span class="n">outfile</span><span class="p">,</span>
                                           <span class="n">stderr</span><span class="o">=</span><span class="n">outfile</span><span class="p">,</span>
                                           <span class="n">close_fds</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
            <span class="n">alf_process</span><span class="o">.</span><span class="n">wait</span><span class="p">()</span>
        <span class="k">if</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">isfile</span><span class="p">(</span><span class="n">infile</span><span class="p">):</span>
            <span class="k">break</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">LOG</span><span class="o">.</span><span class="n">error</span><span class="p">(</span><span class="s1">&#39;ALFSIM was not able to produce simulated sequence.&#39;</span><span class="p">)</span>
        <span class="n">LOG</span><span class="o">.</span><span class="n">error</span><span class="p">(</span><span class="s1">&#39;Do you have ALFSIM installed and used the correct path in the configuration file?&#39;</span><span class="p">)</span>
        <span class="k">return</span>

    <span class="c1"># shutil.rmtree(&#39;/cluster/home/infk/eschaper/spielwiese/&#39;)</span>
    <span class="c1">#shutil.copytree(working_dir, &#39;/cluster/home/infk/eschaper/spielwiese/&#39;)</span>

    <span class="sd">&quot;&quot;&quot; Read in MSA data from ALF</span>
<span class="sd">        The following is a short parser for ALFsim output files</span>
<span class="sd">        yielding MSAs as lists of strings,</span>
<span class="sd">        based on a special flavour of  Felstein stein MSA files &quot;&quot;&quot;</span>

    <span class="c1"># find a repeat unit</span>
    <span class="n">pattern_start</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">compile</span><span class="p">(</span><span class="sa">r</span><span class="s2">&quot;\d+ \d+&quot;</span><span class="p">)</span>
    <span class="n">pattern_seq</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">compile</span><span class="p">(</span><span class="sa">r</span><span class="s2">&quot;\S+[ ]+([A-Z\-]+)&quot;</span><span class="p">)</span>

    <span class="c1"># Our possible parser states:</span>
    <span class="c1">#</span>
    <span class="c1"># state 1: Find beginning of MSA (Felsenstein)</span>
    <span class="c1"># state 2: Find all repeat units</span>

    <span class="c1"># protein ::=</span>
    <span class="c1">#    \d \d</span>
    <span class="c1">#    \S/\S    repeatunits</span>
    <span class="c1">#</span>

    <span class="n">state</span> <span class="o">=</span> <span class="mi">1</span>
    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">infile</span><span class="p">,</span> <span class="s2">&quot;r&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">infile</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">line</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">infile</span><span class="p">):</span>
            <span class="n">LOG</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="s2">&quot;Line </span><span class="si">%d</span><span class="s2">: </span><span class="si">%s</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">i</span><span class="p">,</span> <span class="n">line</span><span class="p">[</span><span class="mi">0</span><span class="p">:</span><span class="o">-</span><span class="mi">1</span><span class="p">])</span>

            <span class="k">if</span> <span class="mi">1</span> <span class="o">==</span> <span class="n">state</span><span class="p">:</span>  <span class="c1"># Find first repeat unit &amp; save begin</span>
                <span class="n">search</span> <span class="o">=</span> <span class="n">pattern_start</span><span class="o">.</span><span class="n">search</span><span class="p">(</span><span class="n">line</span><span class="p">)</span>
                <span class="k">if</span> <span class="n">search</span><span class="p">:</span>
                    <span class="n">LOG</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="s2">&quot; *(1-&gt;2) Found MSA start&quot;</span><span class="p">)</span>
                    <span class="n">state</span> <span class="o">=</span> <span class="mi">2</span>
                    <span class="n">msa</span> <span class="o">=</span> <span class="p">[]</span>

            <span class="k">elif</span> <span class="mi">2</span> <span class="o">==</span> <span class="n">state</span><span class="p">:</span>  <span class="c1"># Find all repeat units</span>
                <span class="n">search</span> <span class="o">=</span> <span class="n">pattern_seq</span><span class="o">.</span><span class="n">search</span><span class="p">(</span><span class="n">line</span><span class="p">)</span>
                <span class="k">if</span> <span class="n">search</span><span class="p">:</span>
                    <span class="n">LOG</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="s2">&quot; *(2-&gt;2) Found another repeat unit&quot;</span><span class="p">)</span>
                    <span class="n">msa</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">search</span><span class="o">.</span><span class="n">group</span><span class="p">(</span><span class="mi">1</span><span class="p">))</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="n">LOG</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="s2">&quot; *(2-&gt;1) repeat region finished, yielding.&quot;</span><span class="p">)</span>
                    <span class="n">state</span> <span class="o">=</span> <span class="mi">1</span>
                    <span class="c1"># YIELD IF WE HAVE FOUND AT LEAST TWO REPEAT UNITS:</span>
                    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">msa</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">1</span><span class="p">:</span>
                        <span class="k">if</span> <span class="n">return_type</span> <span class="o">==</span> <span class="s1">&#39;repeat&#39;</span><span class="p">:</span>
                            <span class="k">yield</span> <span class="n">repeat</span><span class="o">.</span><span class="n">Repeat</span><span class="p">(</span><span class="n">begin</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">msa</span><span class="o">=</span><span class="n">msa</span><span class="p">,</span> <span class="n">sequence_type</span><span class="o">=</span><span class="n">sequence_type</span><span class="p">)</span>
                        <span class="k">elif</span> <span class="n">return_type</span> <span class="o">==</span> <span class="s1">&#39;list&#39;</span><span class="p">:</span>
                            <span class="k">yield</span> <span class="n">msa</span>
                        <span class="k">else</span><span class="p">:</span>
                            <span class="n">LOG</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="s2">&quot;YIELD: </span><span class="si">%s</span><span class="s2">&quot;</span><span class="p">,</span>
                                      <span class="s2">&quot;&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">msa</span><span class="p">)</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s1">&#39;-&#39;</span><span class="p">,</span> <span class="s1">&#39;&#39;</span><span class="p">))</span>
                            <span class="k">yield</span> <span class="n">Bio</span><span class="o">.</span><span class="n">Seq</span><span class="o">.</span><span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">msa</span><span class="p">)</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s1">&#39;-&#39;</span><span class="p">,</span> <span class="s1">&#39;&#39;</span><span class="p">),</span> <span class="n">sequence_type</span><span class="p">)</span>

    <span class="c1"># delete temporary directory</span>
    <span class="k">try</span><span class="p">:</span>
        <span class="n">shutil</span><span class="o">.</span><span class="n">rmtree</span><span class="p">(</span><span class="n">working_dir</span><span class="p">)</span>
    <span class="k">except</span> <span class="ne">OSError</span><span class="p">:</span>
        <span class="k">pass</span></div>


<div class="viewcode-block" id="random_sequence"><a class="viewcode-back" href="../../../tral/repeat.html#tral.repeat.repeat_io.random_sequence">[docs]</a><span class="k">def</span> <span class="nf">random_sequence</span><span class="p">(</span><span class="n">n_samples</span><span class="p">,</span> <span class="n">sequence_type</span><span class="o">=</span><span class="s1">&#39;AA&#39;</span><span class="p">,</span> <span class="n">return_type</span><span class="o">=</span><span class="s1">&#39;repeat&#39;</span><span class="p">,</span>
                    <span class="n">equilibrium_frequencies</span><span class="o">=</span><span class="s1">&#39;human&#39;</span><span class="p">,</span> <span class="n">l</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">n</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span>
                    <span class="n">sequence_length</span><span class="o">=</span><span class="mi">0</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot; Simulate random sequence locally.</span>

<span class="sd">    Simulate random sequence locally.</span>

<span class="sd">    Args:</span>
<span class="sd">        n_samples (int):  The number of samples</span>
<span class="sd">        sequence_type (str): Either &quot;AA&quot; or &quot;DNA&quot;</span>
<span class="sd">        return_type (str): Either &quot;repeat&quot; or &quot;list&quot;</span>
<span class="sd">        equilibrium_frequencies (str): Only &quot;human&quot; option available at current</span>
<span class="sd">        l (int): The length of the repeat unit</span>
<span class="sd">        n (int): The number of repeat units in the tandem repeat</span>
<span class="sd">        sequence_length (int): The total length of the simulated sequence</span>

<span class="sd">    Returns:</span>
<span class="sd">        Return type depends on ``return_type``.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">if</span> <span class="n">sequence_length</span> <span class="o">==</span> <span class="mi">0</span> <span class="ow">and</span> <span class="p">(</span><span class="n">l</span> <span class="o">==</span> <span class="mi">0</span> <span class="ow">or</span> <span class="n">n</span> <span class="o">==</span> <span class="mi">0</span><span class="p">):</span>
        <span class="n">LOG</span><span class="o">.</span><span class="n">error</span><span class="p">(</span><span class="s1">&#39;The specified sequence_length or the product of l and n was&#39;</span>
                  <span class="s1">&#39; set to 0 for random_sequence simulation&#39;</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">sequence_length</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
            <span class="n">sequence_length</span> <span class="o">=</span> <span class="n">l</span> <span class="o">*</span> <span class="n">n</span>

        <span class="c1"># assert equilibrium_frequencies == &#39;human&#39;</span>
        <span class="n">file</span> <span class="o">=</span> <span class="n">config_file</span><span class="p">(</span><span class="s1">&#39;data&#39;</span><span class="p">,</span> <span class="s1">&#39;Random&#39;</span><span class="p">,</span> <span class="s2">&quot;_&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span>
            <span class="p">[</span><span class="n">sequence_type</span><span class="p">,</span> <span class="n">equilibrium_frequencies</span><span class="p">,</span> <span class="s1">&#39;3&#39;</span><span class="p">])</span> <span class="o">+</span> <span class="s1">&#39;.txt&#39;</span><span class="p">)</span>

        <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">file</span><span class="p">,</span> <span class="s1">&#39;r&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
            <span class="n">a</span> <span class="o">=</span> <span class="n">f</span><span class="o">.</span><span class="n">readline</span><span class="p">()[:</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>
        <span class="n">b</span> <span class="o">=</span> <span class="p">[</span><span class="n">i</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39; &#39;</span><span class="p">)</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">a</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;  &#39;</span><span class="p">)]</span>
        <span class="n">frequencies</span> <span class="o">=</span> <span class="p">{</span><span class="n">i</span><span class="p">[</span><span class="mi">0</span><span class="p">]:</span> <span class="nb">int</span><span class="p">(</span><span class="n">i</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">b</span><span class="p">}</span>
        <span class="n">alphabet</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">unique</span><span class="p">([</span><span class="n">i</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">frequencies</span><span class="o">.</span><span class="n">keys</span><span class="p">()])</span>

        <span class="k">for</span> <span class="n">_</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">n_samples</span><span class="p">):</span>
            <span class="n">seed_int</span> <span class="o">=</span> <span class="n">random</span><span class="o">.</span><span class="n">randint</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="nb">sum</span><span class="p">(</span><span class="n">frequencies</span><span class="o">.</span><span class="n">values</span><span class="p">()))</span>
            <span class="k">for</span> <span class="n">key</span><span class="p">,</span> <span class="n">value</span> <span class="ow">in</span> <span class="n">frequencies</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
                <span class="n">seed_int</span> <span class="o">-=</span> <span class="n">value</span>
                <span class="k">if</span> <span class="n">seed_int</span> <span class="o">&lt;=</span> <span class="mi">0</span><span class="p">:</span>
                    <span class="n">seed</span> <span class="o">=</span> <span class="n">key</span>
                    <span class="k">break</span>

            <span class="n">dimer</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">i</span><span class="p">)</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">itertools</span><span class="o">.</span><span class="n">product</span><span class="p">(</span><span class="n">alphabet</span><span class="p">,</span> <span class="n">repeat</span><span class="o">=</span><span class="mi">2</span><span class="p">)]</span>
            <span class="n">third_letter_frequencies</span> <span class="o">=</span> <span class="p">{</span>
                <span class="n">iD</span><span class="p">:</span> <span class="p">{</span>
                    <span class="n">iA</span><span class="p">:</span> <span class="n">frequencies</span><span class="p">[</span>
                        <span class="n">iD</span> <span class="o">+</span>
                        <span class="n">iA</span><span class="p">]</span> <span class="k">for</span> <span class="n">iA</span> <span class="ow">in</span> <span class="n">alphabet</span><span class="p">}</span> <span class="k">for</span> <span class="n">iD</span> <span class="ow">in</span> <span class="n">dimer</span><span class="p">}</span>

            <span class="n">sequence</span> <span class="o">=</span> <span class="n">seed</span>
            <span class="k">for</span> <span class="n">_</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">sequence_length</span> <span class="o">-</span> <span class="mi">3</span><span class="p">):</span>
                <span class="n">next_int</span> <span class="o">=</span> <span class="n">random</span><span class="o">.</span><span class="n">randint</span><span class="p">(</span>
                    <span class="mi">1</span><span class="p">,</span> <span class="nb">sum</span><span class="p">(</span><span class="n">third_letter_frequencies</span><span class="p">[</span><span class="n">sequence</span><span class="p">[</span><span class="o">-</span><span class="mi">2</span><span class="p">:]]</span><span class="o">.</span><span class="n">values</span><span class="p">()))</span>
                <span class="k">for</span> <span class="n">key</span><span class="p">,</span> <span class="n">value</span> <span class="ow">in</span> <span class="n">third_letter_frequencies</span><span class="p">[</span>
                        <span class="n">sequence</span><span class="p">[</span><span class="o">-</span><span class="mi">2</span><span class="p">:]]</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
                    <span class="n">next_int</span> <span class="o">-=</span> <span class="n">value</span>
                    <span class="k">if</span> <span class="n">next_int</span> <span class="o">&lt;=</span> <span class="mi">0</span><span class="p">:</span>
                        <span class="n">sequence</span> <span class="o">+=</span> <span class="n">key</span>
                        <span class="k">break</span>

            <span class="c1"># return Repeat instances</span>
            <span class="k">if</span> <span class="n">return_type</span> <span class="o">==</span> <span class="s1">&#39;repeat&#39;</span> <span class="ow">and</span> <span class="ow">not</span> <span class="n">l</span> <span class="o">==</span> <span class="mi">0</span> <span class="ow">and</span> <span class="ow">not</span> <span class="n">n</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
                <span class="k">yield</span> <span class="p">[</span><span class="n">sequence</span><span class="p">[</span><span class="n">i</span> <span class="o">*</span> <span class="n">l</span><span class="p">:(</span><span class="n">i</span> <span class="o">+</span> <span class="mi">1</span><span class="p">)</span> <span class="o">*</span> <span class="n">l</span><span class="p">]</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">n</span><span class="p">)],</span> <span class="n">sequence_type</span>
            <span class="k">elif</span> <span class="n">return_type</span> <span class="o">==</span> <span class="s1">&#39;list&#39;</span><span class="p">:</span>
                <span class="k">yield</span> <span class="n">sequence</span>
            <span class="k">else</span><span class="p">:</span>  <span class="c1"># return seqIO instances</span>
                <span class="k">yield</span> <span class="n">Bio</span><span class="o">.</span><span class="n">Seq</span><span class="o">.</span><span class="n">Seq</span><span class="p">(</span><span class="n">sequence</span><span class="p">,</span> <span class="n">sequence_type</span><span class="p">)</span></div>
</pre></div>

            <div class="clearer"></div>
          </div>
        </div>
      </div>
      <div class="clearer"></div>
    </div>


  </body>
</html>