
<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>Run and parse de novo repeat detection software. &#8212; TRAL 2.0.0 documentation</title>
    <link rel="stylesheet" href="_static/tral.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <script id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
    <script src="_static/jquery.js"></script>
    <script src="_static/underscore.js"></script>
    <script src="_static/doctools.js"></script>
    <script src="_static/language_data.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Background" href="background.html" />
    <link rel="prev" title="Contributors &amp; Background" href="contributors.html" />
    <link href='http://fonts.googleapis.com/css?family=Source+Sans+Pro:300,400,700|Source+Code+Pro|Armata|IM+Fell+English'
          rel='stylesheet' type='text/css' />
 
    <style type="text/css">
      table.right { float: right; margin-left: 20px; }
      table.right td { border: 1px solid #ccc; }
      
    </style>
    <script type="text/javascript">
      // intelligent scrolling of the sidebar content
      $(window).scroll(function() {
        var sb = $('.sphinxsidebarwrapper');
        var win = $(window);
        var sbh = sb.height();
        var offset = $('.sphinxsidebar').position()['top'];
        var wintop = win.scrollTop();
        var winbot = wintop + win.innerHeight();
        var curtop = sb.position()['top'];
        var curbot = curtop + sbh;
        // does sidebar fit in window?
        if (sbh < win.innerHeight()) {
          // yes: easy case -- always keep at the top
          sb.css('top', $u.min([$u.max([0, wintop - offset - 10]),
                                $(document).height() - sbh - 200]));
        } else {
          // no: only scroll if top/bottom edge of sidebar is at
          // top/bottom edge of window
          if (curtop > wintop && curbot > winbot) {
            sb.css('top', $u.max([wintop - offset - 10, 0]));
          } else if (curtop < wintop && curbot < winbot) {
            sb.css('top', $u.min([winbot - sbh - offset - 20,
                                  $(document).height() - sbh - 200]));
          }
        }
      });
    </script>

  </head><body>
<div class="pageheader">
    <a href="http://www.isb-sib.ch/" class="sib_logo" title="SIB Swiss Institute of Bioinformatics"><img src="_static/sib.png" alt="SIB" class="sib_logo" /></a>
    <a href="index.html">
      <img class="headlogo" src="_static/trallogo.png" alt="TRAL" />
      <span class="headlogo">Tandem Repeat Annotation Library</span>
    </a>
    <div class="headlinks">
        <a href="index.html">Home</a>
        <a href="install.html">Install</a>
        <a href="code_docs.html">Code docs</a>
    </div>
</div>


      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
  <h3><a href="index.html">Table of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">Run and parse de novo repeat detection software.</a><ul>
<li><a class="reference internal" href="#read-in-your-sequences">Read in your sequences.</a></li>
<li><a class="reference internal" href="#run-the-external-repeat-detector">Run the external repeat detector.</a></li>
<li><a class="reference internal" href="#realign-a-repeat-with-propip">Realign a repeat with proPIP.</a></li>
<li><a class="reference internal" href="#output-the-detected-tandem-repeats">Output the detected tandem repeats.</a></li>
</ul>
</li>
</ul>

  <h4>Previous topic</h4>
  <p class="topless"><a href="contributors.html"
                        title="previous chapter">Contributors &amp; Background</a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="background.html"
                        title="next chapter">Background</a></p>
  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="_sources/denovo.rst.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div>
<div id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" />
      <input type="submit" value="Go" />
    </form>
    </div>
</div>
<script>$('#searchbox').show(0);</script>
        </div>
      </div>

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="section" id="run-and-parse-de-novo-repeat-detection-software">
<span id="denovo"></span><h1>Run and parse de novo repeat detection software.<a class="headerlink" href="#run-and-parse-de-novo-repeat-detection-software" title="Permalink to this headline">¶</a></h1>
<p>Here you learn how to run external repeat detection software on your favourite sequence
file and output the results.</p>
<p>Requirements for this tutorial:</p>
<ul class="simple">
<li><p><a class="reference internal" href="install.html#install"><span class="std std-ref">Install TRAL</span></a>. TRAL ships with the data needed for this tutorial.</p></li>
<li><p><a class="reference internal" href="install_external.html#xstream"><span class="std std-ref">Install XSTREAM</span></a>. You can also install one or more other <a class="reference internal" href="install_external.html#install-denovo"><span class="std std-ref">tandem repeat detectors</span></a> instead.</p></li>
<li><p><a class="reference external" href="https://bitbucket.org/lorenzogatti89/castor/wiki/Home">Install Castor</a>. If you wish to realign your repeat MSA you need to have proPIP installed beforehand.</p></li>
</ul>
<div class="admonition-todo admonition" id="id1">
<p class="admonition-title">Todo</p>
<p>write section to install castor with aligner proPIP</p>
</div>
<div class="section" id="read-in-your-sequences">
<h2>Read in your sequences.<a class="headerlink" href="#read-in-your-sequences" title="Permalink to this headline">¶</a></h2>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">os</span>
<span class="kn">from</span> <span class="nn">tral.sequence</span> <span class="kn">import</span> <span class="n">sequence</span>
<span class="kn">from</span> <span class="nn">tral.paths</span> <span class="kn">import</span> <span class="n">PACKAGE_DIRECTORY</span>

<span class="n">proteome_HIV</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">PACKAGE_DIRECTORY</span><span class="p">,</span> <span class="s2">&quot;examples&quot;</span><span class="p">,</span> <span class="s2">&quot;data&quot;</span><span class="p">,</span> <span class="s2">&quot;HIV-1_388796.faa&quot;</span><span class="p">)</span>
<span class="n">sequences_HIV</span> <span class="o">=</span> <span class="n">sequence</span><span class="o">.</span><span class="n">Sequence</span><span class="o">.</span><span class="n">create</span><span class="p">(</span><span class="n">file</span> <span class="o">=</span> <span class="n">proteome_HIV</span><span class="p">,</span> <span class="n">input_format</span> <span class="o">=</span> <span class="s1">&#39;fasta&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="section" id="run-the-external-repeat-detector">
<h2>Run the external repeat detector.<a class="headerlink" href="#run-the-external-repeat-detector" title="Permalink to this headline">¶</a></h2>
<p>Detect tandem repeats on a single sequence with <a class="reference internal" href="install_external.html#xstream"><span class="std std-ref">XSTREAM</span></a>. If you did not
install XSTREAM, you can use any of the other <em>de novo</em> detection algorithms, but will see
different results.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">tandem_repeats</span> <span class="o">=</span> <span class="n">sequences_HIV</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">detect</span><span class="p">(</span><span class="n">denovo</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span> <span class="n">detection</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;detectors&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;XSTREAM&quot;</span><span class="p">]})</span>
</pre></div>
</div>
<p>As an example, the first detected putative tandem repeat looks as follows <a class="reference internal" href="background.html#background"><span class="std std-ref">(interpretation)</span></a>:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">tandem_repeats</span><span class="o">.</span><span class="n">repeats</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
<span class="go">&gt; begin:316 l_effective:4 n:2</span>
<span class="go">GDII</span>
<span class="go">GDIR</span>
</pre></div>
</div>
<p>Detect tandem repeats on all sequences with all <em>de novo</em> tandem repeat detection algorithms
defined in the <a class="reference internal" href="configure.html#configure"><span class="std std-ref">configuration file</span></a>. Here “denovo” is used as a tag, you can choose your own tag.:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="k">for</span> <span class="n">iSequence</span> <span class="ow">in</span> <span class="n">sequences_HIV</span><span class="p">:</span>
    <span class="n">iTandem_repeats</span> <span class="o">=</span> <span class="n">iSequence</span><span class="o">.</span><span class="n">detect</span><span class="p">(</span><span class="n">denovo</span> <span class="o">=</span> <span class="kc">True</span><span class="p">)</span>
    <span class="n">iSequence</span><span class="o">.</span><span class="n">set_repeatlist</span><span class="p">(</span><span class="n">iTandem_repeats</span><span class="p">,</span> <span class="s2">&quot;denovo&quot;</span><span class="p">)</span>
</pre></div>
</div>
<p>Different different algorithms usually detect different tandem repeats. This the the
absolute number of detections in the HIV proteome for a couple of algorithms:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="nb">len</span><span class="p">([</span><span class="n">i</span> <span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="n">sequences_HIV</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">j</span><span class="o">.</span><span class="n">get_repeatlist</span><span class="p">(</span><span class="s1">&#39;denovo&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">repeats</span> <span class="k">if</span> <span class="n">i</span><span class="o">.</span><span class="n">TRD</span> <span class="o">==</span> <span class="s2">&quot;HHrepID&quot;</span><span class="p">])</span>
<span class="go">9</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">len</span><span class="p">([</span><span class="n">i</span> <span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="n">sequences_HIV</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">j</span><span class="o">.</span><span class="n">get_repeatlist</span><span class="p">(</span><span class="s1">&#39;denovo&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">repeats</span> <span class="k">if</span> <span class="n">i</span><span class="o">.</span><span class="n">TRD</span> <span class="o">==</span> <span class="s2">&quot;T-REKS&quot;</span><span class="p">])</span>
<span class="go">4</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">len</span><span class="p">([</span><span class="n">i</span> <span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="n">sequences_HIV</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">j</span><span class="o">.</span><span class="n">get_repeatlist</span><span class="p">(</span><span class="s1">&#39;denovo&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">repeats</span> <span class="k">if</span> <span class="n">i</span><span class="o">.</span><span class="n">TRD</span> <span class="o">==</span> <span class="s2">&quot;TRUST&quot;</span><span class="p">])</span>
<span class="go">6</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">len</span><span class="p">([</span><span class="n">i</span> <span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="n">sequences_HIV</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">j</span><span class="o">.</span><span class="n">get_repeatlist</span><span class="p">(</span><span class="s1">&#39;denovo&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">repeats</span> <span class="k">if</span> <span class="n">i</span><span class="o">.</span><span class="n">TRD</span> <span class="o">==</span> <span class="s2">&quot;XSTREAM&quot;</span><span class="p">])</span>
<span class="go">9</span>
</pre></div>
</div>
<p>As an example, T-REKS detects the following repeat in the second HIV sequence <a class="reference internal" href="background.html#background"><span class="std std-ref">(interpretation)</span></a>:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">([</span><span class="n">i</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">sequences_HIV</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">get_repeatlist</span><span class="p">(</span><span class="s1">&#39;denovo&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">repeats</span> <span class="k">if</span> <span class="n">i</span><span class="o">.</span><span class="n">TRD</span> <span class="o">==</span> <span class="s2">&quot;T-REKS&quot;</span><span class="p">][</span><span class="mi">0</span><span class="p">])</span>
<span class="go">&gt; begin:449 l_effective:10 n:2</span>
<span class="go">RPEPTAPP-ESL</span>
<span class="go">RPEPTAPPPES-</span>
</pre></div>
</div>
</div>
<div class="section" id="realign-a-repeat-with-propip">
<h2>Realign a repeat with proPIP.<a class="headerlink" href="#realign-a-repeat-with-propip" title="Permalink to this headline">¶</a></h2>
<p>Realigning a tandem repeat with an indel aware algortithm such as proPIP may improve its representation.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">test_sequence</span> <span class="o">=</span> <span class="n">sequences_HIV</span><span class="p">[</span><span class="mi">3</span><span class="p">]</span>
<span class="n">tandem_repeats</span> <span class="o">=</span> <span class="n">test_sequence</span><span class="o">.</span><span class="n">detect</span><span class="p">(</span><span class="n">denovo</span> <span class="o">=</span> <span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
<p>Print one of the previous multiple sequence alignments (MSA):</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">tandem_repeats</span><span class="o">.</span><span class="n">repeats</span><span class="p">[</span><span class="mi">6</span><span class="p">])</span>
<span class="go">&gt; begin:828 l_effective:13 n:3</span>
<span class="go">----DKWTVQPIQLPE</span>
<span class="go">---KDSWTVNDIQ--K</span>
<span class="go">LVGKLNWASQIY--PG</span>
</pre></div>
</div>
<p>Realignment with proPIP (default realignment=’proPIP’) of the tandem repeat MSAs with a constant rate distribution:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="k">for</span> <span class="n">TR</span> <span class="ow">in</span> <span class="n">tandem_repeats</span><span class="o">.</span><span class="n">repeats</span><span class="p">:</span>
        <span class="n">TR</span><span class="o">.</span><span class="n">realign_TR</span><span class="p">(</span><span class="n">realignment</span><span class="o">=</span><span class="s1">&#39;proPIP&#39;</span><span class="p">,</span> <span class="n">rate_distribution</span><span class="o">=</span><span class="s1">&#39;constant&#39;</span><span class="p">)</span>
</pre></div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">tandem_repeats</span><span class="o">.</span><span class="n">repeats</span><span class="p">[</span><span class="mi">6</span><span class="p">])</span>
<span class="go">&gt; begin:828 l_effective:13 n:3</span>
<span class="go">---DK--WTVQPIQLPE</span>
<span class="go">--KDS--WTVNDIQ--K</span>
<span class="go">LV-GKLNWASQ-I-YPG</span>
</pre></div>
</div>
<p>Set rate_distribution = ‘gamma’ for a gamma distributed indel placement or modify it in <a class="reference internal" href="configure.html#configure"><span class="std std-ref">config.ini</span></a>.
Set realignment = ‘mafft’ to realign with mafft instead of proPIP.</p>
</div>
<div class="section" id="output-the-detected-tandem-repeats">
<h2>Output the detected tandem repeats.<a class="headerlink" href="#output-the-detected-tandem-repeats" title="Permalink to this headline">¶</a></h2>
<p>Write a singe repeat_list to .tsv format:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">path_to_output_tsv_file</span> <span class="o">=</span> <span class="s2">&quot;outputfile.tsv&quot;</span> <span class="c1"># Choose your path and filename</span>
<span class="n">tandem_repeats</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">output_format</span> <span class="o">=</span> <span class="s2">&quot;tsv&quot;</span><span class="p">,</span> <span class="n">file</span> <span class="o">=</span> <span class="n">path_to_output_tsv_file</span><span class="p">)</span>
</pre></div>
</div>
<p>The created .tsv looks as follows <a class="reference internal" href="background.html#background"><span class="std std-ref">(interpretation)</span></a>:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ cat outputfile.tsv
begin   msa_original    l_effective      nD      sequence_length divergence      pvalue
316     GDII,GDIR       4       2.0     8       None    None
507     FLG,FLG 3       2.0     6       None    None
</pre></div>
</div>
<p>Write a singe repeat_list to .pickle format:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">path_to_output_pickle_file</span> <span class="o">=</span> <span class="s2">&quot;outputfile.pickle&quot;</span>  <span class="c1"># Choose your path and filename</span>
<span class="n">tandem_repeats</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">output_format</span> <span class="o">=</span> <span class="s2">&quot;pickle&quot;</span><span class="p">,</span> <span class="n">file</span> <span class="o">=</span> <span class="n">path_to_output_pickle_file</span><span class="p">)</span>
</pre></div>
</div>
<p>A repeat_list in pickle format can easily be read in again:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">tral.repeat_list</span> <span class="kn">import</span> <span class="n">repeat_list</span>
<span class="n">tandem_repeats</span> <span class="o">=</span> <span class="n">repeat_list</span><span class="o">.</span><span class="n">RepeatList</span><span class="o">.</span><span class="n">create</span><span class="p">(</span><span class="n">input_format</span> <span class="o">=</span> <span class="s2">&quot;pickle&quot;</span><span class="p">,</span> <span class="n">file</span> <span class="o">=</span> <span class="n">path_to_output_pickle_file</span><span class="p">)</span>
</pre></div>
</div>
<p>Save multiple sequence together with tandem repeat annotations:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">pickle</span>
<span class="n">path_to_output_pickle_file</span> <span class="o">=</span> <span class="s2">&quot;outputfile.pickle&quot;</span> <span class="c1"># Choose your path and filename</span>
<span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">path_to_output_pickle_file</span><span class="p">,</span> <span class="s1">&#39;wb&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">fh</span><span class="p">:</span>
    <span class="n">pickle</span><span class="o">.</span><span class="n">dump</span><span class="p">(</span><span class="n">sequences_HIV</span><span class="p">,</span> <span class="n">fh</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>


            <div class="clearer"></div>
          </div>
        </div>
      </div>
      <div class="clearer"></div>
    </div>


  </body>
</html>