
<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>Use GC3Pie to annotate your large sequence dataset. &#8212; TRAL 2.0.0 documentation</title>
    <link rel="stylesheet" href="_static/tral.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <script id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
    <script src="_static/jquery.js"></script>
    <script src="_static/underscore.js"></script>
    <script src="_static/doctools.js"></script>
    <script src="_static/language_data.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Searching for particular repeats" href="search_hmm.html" />
    <link rel="prev" title="Perform overlap filtering of redundant tandem repeat annotations." href="overlap_filtering.html" />
    <link href='http://fonts.googleapis.com/css?family=Source+Sans+Pro:300,400,700|Source+Code+Pro|Armata|IM+Fell+English'
          rel='stylesheet' type='text/css' />
 
    <style type="text/css">
      table.right { float: right; margin-left: 20px; }
      table.right td { border: 1px solid #ccc; }
      
    </style>
    <script type="text/javascript">
      // intelligent scrolling of the sidebar content
      $(window).scroll(function() {
        var sb = $('.sphinxsidebarwrapper');
        var win = $(window);
        var sbh = sb.height();
        var offset = $('.sphinxsidebar').position()['top'];
        var wintop = win.scrollTop();
        var winbot = wintop + win.innerHeight();
        var curtop = sb.position()['top'];
        var curbot = curtop + sbh;
        // does sidebar fit in window?
        if (sbh < win.innerHeight()) {
          // yes: easy case -- always keep at the top
          sb.css('top', $u.min([$u.max([0, wintop - offset - 10]),
                                $(document).height() - sbh - 200]));
        } else {
          // no: only scroll if top/bottom edge of sidebar is at
          // top/bottom edge of window
          if (curtop > wintop && curbot > winbot) {
            sb.css('top', $u.max([wintop - offset - 10, 0]));
          } else if (curtop < wintop && curbot < winbot) {
            sb.css('top', $u.min([winbot - sbh - offset - 20,
                                  $(document).height() - sbh - 200]));
          }
        }
      });
    </script>

  </head><body>
<div class="pageheader">
    <a href="http://www.isb-sib.ch/" class="sib_logo" title="SIB Swiss Institute of Bioinformatics"><img src="_static/sib.png" alt="SIB" class="sib_logo" /></a>
    <a href="index.html">
      <img class="headlogo" src="_static/trallogo.png" alt="TRAL" />
      <span class="headlogo">Tandem Repeat Annotation Library</span>
    </a>
    <div class="headlinks">
        <a href="index.html">Home</a>
        <a href="install.html">Install</a>
        <a href="code_docs.html">Code docs</a>
    </div>
</div>


      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
  <h3><a href="index.html">Table of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">Use GC3Pie to annotate your large sequence dataset.</a><ul>
<li><a class="reference internal" href="#configure-tral">Configure TRAL.</a></li>
<li><a class="reference internal" href="#prepare-your-data">Prepare your data.</a><ul>
<li><a class="reference internal" href="#sequence-data">Sequence data</a></li>
<li><a class="reference internal" href="#sequence-profile-model-to-sequence-data-mapping">Sequence profile model to sequence data mapping</a></li>
<li><a class="reference internal" href="#sequence-profile-models">Sequence profile models</a></li>
<li><a class="reference internal" href="#pyfaidx">pyfaidx</a></li>
<li><a class="reference internal" href="#test-the-file-preprocessing">Test the file preprocessing.</a></li>
</ul>
</li>
<li><a class="reference internal" href="#perform-a-test-run">Perform a test run.</a></li>
<li><a class="reference internal" href="#install-gc3pie">Install GC3PIE.</a></li>
<li><a class="reference internal" href="#usage">Usage</a></li>
</ul>
</li>
</ul>

  <h4>Previous topic</h4>
  <p class="topless"><a href="overlap_filtering.html"
                        title="previous chapter">Perform overlap filtering of redundant tandem repeat annotations.</a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="search_hmm.html"
                        title="next chapter">Searching for particular repeats</a></p>
  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="_sources/workflow.rst.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div>
<div id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" />
      <input type="submit" value="Go" />
    </form>
    </div>
</div>
<script>$('#searchbox').show(0);</script>
        </div>
      </div>

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="section" id="use-gc3pie-to-annotate-your-large-sequence-dataset">
<span id="workflow"></span><h1>Use GC3Pie to annotate your large sequence dataset.<a class="headerlink" href="#use-gc3pie-to-annotate-your-large-sequence-dataset" title="Permalink to this headline">¶</a></h1>
<p>This tutorial will work you through an example of annotating tandem repeats on a very
large sequence data set: We will annotate tandem repeats with sequence profile models,
and with <em>de novo</em> detection algorithms, perform significance testing and overlap
filtering all in one.</p>
<p>For very large sequence sets (&gt; 6000 sequences ~ 10h runtime), you may wish to perform the
annotation on several computing  nodes in parallel. Here, we provide a system to automise
the distribution and collection of tandem repeat annotation jobs with
<a class="reference external" href="https://gc3pie.readthedocs.io/en/master/index.html">GC3Pie</a>.</p>
<p>Everything is explained with a <a class="reference external" href="https://github.com/acg-team/tral/tree/develop/tral/examples/workflow">toy workflow located in the TRAL source</a>. You can print
the path to the local copy included in TRAL with this Python code:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">os</span>
<span class="kn">from</span> <span class="nn">tral.paths</span> <span class="kn">import</span> <span class="n">PACKAGE_DIRECTORY</span>
<span class="nb">print</span><span class="p">(</span><span class="n">PACKAGE_DIRECTORY</span><span class="p">)</span>
</pre></div>
</div>
<p>To follow this tutorial, set a shell variable $MYTRAL to the package variable PACKAGE_DIRECTORY. In bash and sh, this can be done with:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">export</span> <span class="n">MYTRAL</span><span class="o">=/</span><span class="n">path</span><span class="o">/</span><span class="n">to</span><span class="o">/</span><span class="n">your</span><span class="o">/</span><span class="n">package</span><span class="o">/</span><span class="n">directory</span>
</pre></div>
</div>
<p>For csh or tcsh, use:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">setenv</span> <span class="n">MYTRAL</span> <span class="o">/</span><span class="n">path</span><span class="o">/</span><span class="n">to</span><span class="o">/</span><span class="n">your</span><span class="o">/</span><span class="n">package</span><span class="o">/</span><span class="n">directory</span>
</pre></div>
</div>
<div class="section" id="configure-tral">
<h2>Configure TRAL.<a class="headerlink" href="#configure-tral" title="Permalink to this headline">¶</a></h2>
<p>We start by <a class="reference internal" href="configure.html#configure"><span class="std std-ref">adapting the TRAL configuration files</span></a> to choose for
example which external tandem repeat detection algorithms you would like to run on the
sequence data. Don’t forget to add a comma behind the last entry in this list, otherwise the
file format is violated.</p>
<p>Now, let’s move ahead to data acquisition:</p>
</div>
<div class="section" id="prepare-your-data">
<h2>Prepare your data.<a class="headerlink" href="#prepare-your-data" title="Permalink to this headline">¶</a></h2>
<p>To annotate sequences with tandem repeats, we already acquired the following data and stored it in <em>$MYTRAL/examples/workflow</em> :</p>
<div class="section" id="sequence-data">
<h3>Sequence data<a class="headerlink" href="#sequence-data" title="Permalink to this headline">¶</a></h3>
<p>We downloaded all <a class="reference external" href="http://www.uniprot.org/uniprot/?query=gene%3Aprdm+AND+reviewed%3Ayes+AND+organism%3A%22Homo+sapiens+%28Human%29+%5B9606%5D%22&amp;sort=score">UniProtKB/Swiss-Prot human PRDM genes</a>
in fasta format…</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>cd $MYTRAL/examples/workflow
less uniprot_PRDM.fasta
</pre></div>
</div>
<p>… and split the sequence file into subsets on which the tandem repeat annotation can be
performed in parallel.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">ls</span> <span class="n">split_sequence_data</span><span class="o">/*</span>
</pre></div>
</div>
<p>On a sidenote, <a class="reference external" href="ftp://saf.bio.caltech.edu/pub/software/molbio/fastasplitn.c">fastasplitn</a>
is a useful tool for splitting large data files.</p>
</div>
<div class="section" id="sequence-profile-model-to-sequence-data-mapping">
<h3>Sequence profile model to sequence data mapping<a class="headerlink" href="#sequence-profile-model-to-sequence-data-mapping" title="Permalink to this headline">¶</a></h3>
<p>We downloaded the mapping to PFAM annotations for these sequence also from www.uniprot.org.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">less</span> <span class="n">uniprot_PRDM_annotation</span><span class="o">.</span><span class="n">tsv</span>
</pre></div>
</div>
</div>
<div class="section" id="sequence-profile-models">
<h3>Sequence profile models<a class="headerlink" href="#sequence-profile-models" title="Permalink to this headline">¶</a></h3>
<p>We downloaded all needed PFAM sequence profile models, for example of the <a class="reference external" href="http://pfam.xfam.org/family/PF00096/hmm">Zinc finger</a>.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">less</span> <span class="n">PRDM_PFAM_models</span><span class="o">.</span><span class="n">hmm</span>
</pre></div>
</div>
</div>
<div class="section" id="pyfaidx">
<h3>pyfaidx<a class="headerlink" href="#pyfaidx" title="Permalink to this headline">¶</a></h3>
<p>The toy example requires the Python package <a class="reference external" href="https://github.com/mdshw5/pyfaidx">pyfaidx</a>
for .fasta file indexing.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">pip</span> <span class="n">install</span> <span class="n">pyfaidx</span>
</pre></div>
</div>
</div>
<div class="section" id="test-the-file-preprocessing">
<h3>Test the file preprocessing.<a class="headerlink" href="#test-the-file-preprocessing" title="Permalink to this headline">¶</a></h3>
<p>To allow fast access to the data for the tandem repeat annotation steps, it is transformed
to Python pickle objects:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>mkdir $MYTRAL/examples/workflow/hmm
python3 tandem_repeat_annotation_scripts.py file_preparation --hmm_annotation_raw $MYTRAL/examples/workflow/uniprot_PRDM_annotation.tsv --hmm_annotation $MYTRAL/examples/workflow/uniprot_PRDM_annotation.pickle --hmm_raw $MYTRAL/examples/workflow/PRDM_PFAM_models.hmm --hmm $MYTRAL/examples/workflow/hmm
</pre></div>
</div>
<p>On the small toy example, this file preparation step should run fast.</p>
</div>
</div>
<div class="section" id="perform-a-test-run">
<h2>Perform a test run.<a class="headerlink" href="#perform-a-test-run" title="Permalink to this headline">¶</a></h2>
<p>With the following command, annotation is performed on <em>uniprot_PRDM_1.fasta</em>:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>python3 tandem_repeat_annotation_scripts.py workflow -i $MYTRAL/examples/workflow/split_sequence_data/uniprot_PRDM_1.fasta -o $MYTRAL/examples/workflow/results/uniprot_PRDM_1.pickle -os $MYTRAL/examples/workflow/results/uniprot_PRDM_1.tsv -f tsv -t 600  --hmm_annotation $MYTRAL/examples/workflow/uniprot_PRDM_annotation.pickle --hmm $MYTRAL/examples/workflow/hmm
</pre></div>
</div>
<p>If this runs fine, you should see annotation results in:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">less</span> <span class="n">results</span><span class="o">/</span><span class="n">uniprot_PRDM_1</span><span class="o">.</span><span class="n">tsv</span>
</pre></div>
</div>
<p>Now, we can move ahead to automated distributed annotation with GC3PIE.</p>
</div>
<div class="section" id="install-gc3pie">
<h2>Install GC3PIE.<a class="headerlink" href="#install-gc3pie" title="Permalink to this headline">¶</a></h2>
<p>You can follow the official
<a class="reference external" href="https://gc3pie.readthedocs.io/en/master/users/install.html">GC3Pie installation instructions</a>.
Upon installation, GC3Pie needs to be <a class="reference external" href="http://gc3pie.readthedocs.org/en/latest/users/configuration.html">locally configured</a>.</p>
</div>
<div class="section" id="usage">
<h2>Usage<a class="headerlink" href="#usage" title="Permalink to this headline">¶</a></h2>
<p>Adapt this command to run the tandem repeat annotation workflow (<a class="reference external" href="https://gc3pie.readthedocs.io/en/master/users/configuration.html">more information</a>):</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ ./tandem_repeat_annotation_workflow.py -w 60 minutes -r &lt;host&gt; -J 500 -u sqlite:////path/to/&lt;session_name&gt;.db -s &lt;session_name&gt; -C 2 -vvvv -conf $MYTRAL/examples/workflow/tandem_repeat_annotation_workflow.ini
</pre></div>
</div>
<p>Control the workflow with <a class="reference external" href="https://gc3pie.readthedocs.io/en/master/users/gc3utils.html">GC3Utils</a>.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ ./tandemrepeatannotationworkflow -s &lt;session_name&gt;
$ gselect
$ gresub
$ gstat
$ ...
</pre></div>
</div>
</div>
</div>


            <div class="clearer"></div>
          </div>
        </div>
      </div>
      <div class="clearer"></div>
    </div>


  </body>
</html>